
<hr style="border:2px solid gray"> </hr>

# $\color{blue}{\text{1. Introducción a Qiskit}}$
<hr style="border:2px solid gray"> </hr>


**Qiskit (Quantum Information Software Kit)** es una biblioteca de código abierto desarrollada por IBM para programar y trabajar con computación cuántica.




Es una de las herramientas más populares y ampliamente utilizadas en el campo de la computación cuántica. Qiskit proporciona un conjunto de herramientas y librerías que permiten a los desarrolladores y científicos aprovechar las capacidades de los ordenadores cuánticos.

Qiskit está escrito principalmente en Python y proporciona una interfaz de programación de aplicaciones (API) que permite a los usuarios construir y ejecutar circuitos cuánticos, realizar simulaciones, realizar mediciones y analizar los resultados. También ofrece acceso a los ordenadores cuánticos reales a través de la nube de IBM Quantum Experience.

La página de Qiskit con toda la información está en

https://qiskit.org/


La biblioteca Qiskit consta de varios componentes principales:

**Qiskit Terra**: es el núcleo de Qiskit y proporciona las herramientas necesarias para construir circuitos cuánticos y manipularlos. Incluye las clases para la representación de circuitos cuánticos, operaciones cuánticas, puertas cuánticas, medidas, etc.

**Qiskit Aer**: es un paquete que permite simular circuitos cuánticos en ordenadores clásicos. Proporciona simuladores de alta fidelidad para realizar cálculos y experimentos en entornos de simulación controlados.

**Qiskit Aqua**: es una biblioteca de Qiskit especializada en aplicaciones cuánticas de alto nivel, como la optimización cuántica, la química cuántica y el aprendizaje automático cuántico. Proporciona algoritmos y herramientas específicas para resolver problemas en estos dominios.

**Qiskit Ignis**: es un conjunto de herramientas para caracterizar y mitigar errores en los ordenadores cuánticos. Ayuda a entender y corregir los errores inherentes a los sistemas cuánticos reales, mejorando así la calidad y precisión de los resultados obtenidos.

Qiskit se ha convertido en una opción popular tanto para principiantes como para expertos en computación cuántica debido a su facilidad de uso, su enfoque modular y su capacidad para interactuar con los ordenadores cuánticos reales. Además, cuenta con una amplia documentación, tutoriales y una comunidad activa que brinda soporte y comparte conocimientos en el desarrollo de aplicaciones cuánticas.

Más detalles en

https://en.wikipedia.org/wiki/Qiskit








Nótese que qiskit es uno de los muchos lenguajes de programación cuántica que se han desarrollado

https://en.wikipedia.org/wiki/Quantum_programming

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{1.1 Instalación de Qiskit}}$
<hr style="border:2px solid gray"> </hr>



Las instrucciones detalladas para instalar Qiskit están en

https://qiskit.org/documentation/getting_started.html



En nuestro caso que usamos Jupyter Notebook en Google Colab
es suficiente con instalar el paquete *qiskit* y el módulo *pylatexenc* para dibujar los diagramas con matplotlib

Google colab tiene los siguientes paquetes y módulos pre-instalados

In [ ]:
!pip list

Usamos **pip** para instalar nuevos paquetes y módulos.

In [1]:
!pip install qiskit
!pip install pylatexenc # instalamos la librería *pylatexenc* para dibujar circuitos con matplotlib

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.9/5.9 MB 15.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 28.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 241.5/241.5 kB 19.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.9/129.9 kB 16.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 69.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 kB 11.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 5.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.5/37.5 MB 22.4 MB/s eta 0:00:00


In [ ]:
!pip list

In [2]:
import qiskit
qiskit.__qiskit_version__

{'qiskit-terra': '0.24.1', 'qiskit-aer': '0.12.0', 'qiskit-ignis': None, 'qiskit-ibmq-provider': '0.20.2', 'qiskit': '0.43.1', 'qiskit-nature': None, 'qiskit-finance': None, 'qiskit-optimization': None, 'qiskit-machine-learning': None}

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{1.2 Tutoriales de Qiskit}}$
<hr style="border:2px solid gray"> </hr>
Se puede aprender a usar Qiskit a través de los tutoriales
propios, por ejemplo en

https://qiskit.org/documentation/intro_tutorial1.html





<hr style="border:2px solid gray"> </hr>

# $\color{blue}{\text{2. Programación Cuántica usando Qiskit}}$
<hr style="border:2px solid gray"> </hr>

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{2.1 Repaso de Compuertas Cuánticas}}$
<hr style="border:2px solid gray"> </hr>


**Estados cuánticos para 1 qubit** se representan como kets o como vectores columna

Podemos usar diferentes bases para representar el estado cuántico de un
qubit:


Base Z (Base computacional)
$$
|0\rangle = \begin{pmatrix} 1 \\\\ 0 \end{pmatrix} \, \, \, \, |1\rangle =\begin{pmatrix} 0 \\\\ 1 \end{pmatrix}.
$$


Base X
$$
|\pm\rangle = \frac{|0\rangle \pm|1\rangle}{\sqrt{2}}=\frac{1}{\sqrt{2}}\begin{pmatrix} 1 \\\\ \pm1 \end{pmatrix}.
$$


Base Y

$$
|\circlearrowright\rangle = \frac{ | 0 \rangle + i | 1 \rangle}{\sqrt{2}} = \frac{1}{\sqrt{2}}\begin{pmatrix} 1 \\\\ i \end{pmatrix}, ~~~~ |\circlearrowleft\rangle = \frac{ | 0 \rangle -i | 1 \rangle}{\sqrt{2}} = \frac{1}{\sqrt{2}}\begin{pmatrix} 1 \\\\ -i \end{pmatrix}.
$$

Para el estado superpuesto
<div align="center"> $|\psi\rangle= \alpha|0\rangle + \beta|1\rangle$ </div>

la probabilidad de medir $|0\rangle$ es $\alpha \alpha^*$ y la probabilidad de medir $|1\rangle$ es $\beta \beta^*$

En general para un qubit

$$
|\psi\rangle = \cos{\frac{\theta}{2}}|0\rangle + e^{i\phi}\sin{\frac{\theta}{2}}|1\rangle
$$
con $0\leq \phi < 2\pi$, y $0\leq \theta \leq \pi$.

**Los operadores cuánticos son unitarios**

$$\left|\psi'\right\rangle = U\left|\psi\right\rangle$$

$$
U U^\dagger = U^\dagger U = 1.
$$

A $U^\dagger$ se le llama **transpuesto conjugado** y corresponde al inverso de $U$


Las mediciones cuánticas se representan por operadores Hermíticos

$$
H = H^\dagger.
$$

que tienen eigenvalores reales.


Por ejemplo un operador unitario que transforma $\left|0\right\rangle$ en  $\left|\psi\right\rangle$ es

$$
U = \begin{pmatrix}
\cos(\theta/2) & a \\
e^{i\phi}\sin(\theta/2) & b
\end{pmatrix}
$$

con $a$ y $b$ números complejos de modo que  $U^\dagger U = I$ para todos los ángulos $0\leq\theta\leq\pi$ y $0\leq \phi<2\pi$. Con $0\leq \lambda<2\pi$

$$
U(\theta, \phi, \lambda) =
    \begin{pmatrix}
        \cos\left(\frac{\theta}{2}\right)          & -e^{i\lambda}\sin\left(\frac{\theta}{2}\right) \\
        e^{i\phi}\sin\left(\frac{\theta}{2}\right) & e^{i(\phi+\lambda)}\cos\left(\frac{\theta}{2}\right)
    \end{pmatrix}
$$


**Los operadores unitarios son reversibles y no cambian la norma del estado cuántico**.

La medición no es una operación unitaria y tampoco es reversible.


Matrices de Pauli
$$
X= \begin{pmatrix} 0&1 \\\\ 1&0 \end{pmatrix} \hspace{2cm}
Y= \begin{pmatrix} 0&-i \\\\ i&0 \end{pmatrix}\hspace{2cm}
Z= \begin{pmatrix} 1&0 \\\\ 0&-1 \end{pmatrix}
$$


Recordemos que los productos externos de estados de un qubit
se pueden representar como matrices 2x2
$$
|0\rangle\langle0|= \begin{pmatrix} 1 \\\\ 0 \end{pmatrix} \begin{pmatrix} 1 & 0 \end{pmatrix} = \begin{pmatrix} 1&0 \\\\ 0&0 \end{pmatrix},\\\\
|0\rangle\langle1| = \begin{pmatrix} 1 \\\\ 0 \end{pmatrix} \begin{pmatrix} 0 & 1 \end{pmatrix} = \begin{pmatrix} 0&1 \\\\ 0&0 \end{pmatrix},\\\\
|1\rangle\langle0| = \begin{pmatrix} 0 \\\\ 1 \end{pmatrix} \begin{pmatrix} 1 & 0 \end{pmatrix} = \begin{pmatrix} 0&0 \\\\ 1&0 \end{pmatrix},\\\\
|1\rangle\langle1| = \begin{pmatrix} 0 \\\\ 1 \end{pmatrix} \begin{pmatrix} 0 & 1 \end{pmatrix} = \begin{pmatrix} 0&0 \\\\ 0&1 \end{pmatrix}.\\\\
$$


$$\newcommand{\ket}[1]{\left|{#1}\right\rangle}$$
$$\newcommand{\bra}[1]{\left\langle{#1}\right|}$$

Compuerta X de Pauli

$$ X =\ket{0}\bra{1} + \ket{1}\bra{0}$$

$$ X \ket{0} = \ket{1}$$
$$ X \ket{1} = \ket{0}$$

o en forma matricial

$$
X= \begin{pmatrix} 0&1 \\\\ 1&0 \end{pmatrix} \hspace{2cm}
$$

Compuerta Y de Pauli
$$ Y =-i\ket{0}\bra{1} + i\ket{1}\bra{0}$$

$$ Y \ket{0}= i\ket{1}$$
$$ Y \ket{1}= -i\ket{0}$$

o en forma matricial

$$
Y= \begin{pmatrix} 0&-i \\\\ i&0 \end{pmatrix}\hspace{2cm}
$$

Compuerta Z de Pauli
$$ Z =\ket{0}\bra{0} - \ket{1}\bra{1}$$

$$ Z \ket{0}= \ket{0} $$
$$ Z \ket{1}= -\ket{1} $$

o en forma matricial

$$
Z= \begin{pmatrix} 1&0 \\\\ 0&-1 \end{pmatrix}
$$

Compuerta H de Hadamard
$$ H =\frac{\ket{0}+\ket{1}}{\sqrt{2}} \bra{0}+ \frac{\ket{0}-\ket{1}}{\sqrt{2}} \bra{1}$$

$$ H \ket{0} =\frac{\ket{0}+\ket{1}}{\sqrt{2}}$$
$$ H \ket{1} =\frac{\ket{0}-\ket{1}}{\sqrt{2}}$$

o en forma matricial

$$
H= \frac{1}{\sqrt{2}}\begin{pmatrix} 1&1 \\\\ 1&-1 \end{pmatrix}
$$


Compuerta CNOT aplicada a 2 quibits, el primer bit (el que está a la derecha), se llama de control y el segundo se llama objetivo. Si el qubit de control es 1 el qubit objetivo se niega.

$$ CNOT \ket{00}= \ket{00}$$
$$ CNOT \ket{01}= \ket{11}$$
$$ CNOT \ket{10}= \ket{10}$$
$$ CNOT \ket{11}= \ket{01}$$


**Producto Tensorial**

El espacio de una computadora cuántica crece exponencialmente con el número de qubits, para $n$ qubits el espacio vectorial complejo tiene dimensión $d=2^n$. Se usa el **producto tensorial** para describir estados y operadores con multi-qubits.

Para un sistema de 2 qubits y 2 operadores $A$ y $B$ que actúan en su respectivo qubit, el producto tensorial de estados y operadores es

$$\begin{equation}\begin{split}
	\left|{00}\right\rangle &= \begin{pmatrix}
		1 \begin{pmatrix}
			1  \\
			0
		\end{pmatrix} \\
		0 \begin{pmatrix}
			1  \\
			0
		\end{pmatrix}
	\end{pmatrix} = \begin{pmatrix} 1 \\ 0 \\ 0 \\0 \end{pmatrix}~~~\left|{01}\right\rangle = \begin{pmatrix}
	1 \begin{pmatrix}
	0 \\
	1
	\end{pmatrix} \\
	0 \begin{pmatrix}
	0  \\
	1
	\end{pmatrix}
	\end{pmatrix} = \begin{pmatrix}0 \\ 1 \\ 0 \\ 0 \end{pmatrix}\end{split}
\end{equation}$$

$$\begin{equation}\begin{split}\left|{10}\right\rangle = \begin{pmatrix}
	0\begin{pmatrix}
	1  \\
	0
	\end{pmatrix} \\
	1\begin{pmatrix}
	1 \\
	0
	\end{pmatrix}
	\end{pmatrix} = \begin{pmatrix} 0 \\ 0 \\ 1 \\ 0 \end{pmatrix}~~~ 	\left|{11}\right\rangle = \begin{pmatrix}
	0 \begin{pmatrix}
	0  \\
	1
	\end{pmatrix} \\
	1\begin{pmatrix}
	0  \\
	1
	\end{pmatrix}
	\end{pmatrix} = \begin{pmatrix} 0 \\ 0 \\ 0 \\1 \end{pmatrix}\end{split}
\end{equation}.$$



$$\begin{equation}
	A\otimes B =
	\begin{pmatrix}
		A_{00} \begin{pmatrix}
			B_{00} & B_{01} \\
			B_{10} & B_{11}
		\end{pmatrix} & A_{01} 	\begin{pmatrix}
				B_{00} & B_{01} \\
				B_{10} & B_{11}
			\end{pmatrix} \\
		A_{10} 	\begin{pmatrix}
					B_{00} & B_{01} \\
					B_{10} & B_{11}
				\end{pmatrix} & A_{11} 	\begin{pmatrix}
							B_{00} & B_{01} \\
							B_{10} & B_{11}
						\end{pmatrix}
	\end{pmatrix},
\end{equation}$$





Donde $\left|0\right\rangle \otimes \left|0\right\rangle$ se abrevia como $\left|00\right\rangle$.

Por ejemplo

$$
Z \otimes X=  \begin{bmatrix} 1&0 \\ 0&-1 \end{bmatrix} \otimes \begin{bmatrix} 0&1 \\ 1&0 \end{bmatrix} = \begin{bmatrix} 0&1&0&0 \\ 1&0&0&0\\0&0&0&-1\\0&0&-1&0 \end{bmatrix}.
$$




**Compuerta SWAP**

SWAP intercambia los 2 qubits

$$\left|00\right\rangle \rightarrow \left|00\right\rangle~,~\left|01\right\rangle \rightarrow \left|10\right\rangle~,~\left|10\right\rangle \rightarrow \left|01\right\rangle~,~\left|11\right\rangle \rightarrow \left|11\right\rangle,$$

o en forma matricial

$$
\mathrm{SWAP} =
\begin{pmatrix}
1 & 0 & 0 & 0\\
0 & 0 & 1 & 0\\
0 & 1 & 0 & 0\\
0 & 0 & 0 & 1
\end{pmatrix}.
$$

**Estados de 3 qubits**


La base computacional es

$$\left|000\right\rangle, \left|001\right\rangle, \left|010\right\rangle, \left|011\right\rangle, \left|100\right\rangle, \left|101\right\rangle, \left|110\right\rangle, \left|111\right\rangle,$$


$$\left|abc\right\rangle : \underset{\text{qubit 2}}{\left|a\right\rangle}\otimes \underset{\text{qubit 1}}{\left|b\right\rangle}\otimes \underset{\text{qubit 0}}{\left|c\right\rangle}.$$

Para simplificar la notación en algoritmos es común usar la notación

$$\left|0\right\rangle, \left|1\right\rangle, \left|2\right\rangle, \left|3\right\rangle, \left|4\right\rangle, \left|5\right\rangle, \left|6\right\rangle, \left|7\right\rangle,$$


<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{2.2 Un Primer Circuito Cuántico Arbitrario}}$
<hr style="border:2px solid gray"> </hr>

Ahora hacemos un primer circuito cuántico arbitrario en *qiskit*.

*Nota: El estado inicial de los qubits es $\ket{0}$*

In [ ]:
from qiskit import QuantumCircuit, Aer
import numpy as np
import qiskit.quantum_info as qi
qc = QuantumCircuit(4,4) # Ponemos 4 qubits y 4 bits
# Aplicamos algunas compuertas
qc.x(0)
qc.y(1)
qc.z(2)
qc.h(3)
stv = qi.Statevector.from_instruction(qc)
# Ahora hacemos mediciones de cada qubit
qc.measure(0,0)
qc.measure(1,1)
qc.measure(2,2)
qc.measure(3,3)
qc.draw(output='mpl')

Las compuertas cuánticas aplicadas se expresan matemáticamente del modo siguiente:

$$
\begin{align*} (H \otimes Z \otimes Y \otimes X) |0000\rangle =\frac{\ket{0}+\ket{1}}{\sqrt{2}} \otimes|0\rangle\otimes i|1\rangle\otimes |1\rangle = \frac{i}{\sqrt{2}}|0011\rangle+ \frac{i}{\sqrt{2}}|1011\rangle
\end{align*}
$$


El mismo resultado se obtiene con qiskit   para el  estado del sistema antes de las mediciones con la líneas

*stv = qi.Statevector.from_instruction(qc)*

y

*stv.draw('latex', prefix="|\\psi\\rangle =")*



In [97]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

Alternativamente podemos hacer las operaciones en la
aplicacion web

https://ciiec.buap.mx/QCSim/

escribiendo

h q[0];

y q[1];

z q[2];

h q[3];

Enseguida simulamos este circuito usando el *backend qasm_simulator* y mostramos los resultados estadísticos de ejecutarlo muchas veces.

Nota: La notación de *qiskit* pone el qubit más significativo a la izquierda igual que en los números binarios.

In [ ]:
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1000)
print(job.result().get_counts())
plot_histogram(job.result().get_counts())

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{2.3 Circuito Cuántico para Superposición Simétrica}}$
<hr style="border:2px solid gray"> </hr>

Para poner en superposición uniforme a todos los qubits aplicamos la compuerta de Hadamard a todos los qubits

In [ ]:
from qiskit import QuantumCircuit
import qiskit.quantum_info as qi
qc = QuantumCircuit(3,3)
for qubit in range(3):
  qc.h(qubit)
stv = qi.Statevector.from_instruction(qc)
for qubit in range(3):
  qc.measure(qubit,qubit)
qc.draw("mpl")

Las compuertas cuánticas aplicadas se expresan matemáticamente del modo siguiente:

$$
\begin{align*} (H \otimes H \otimes H ) |000\rangle =\frac{\ket{0}+\ket{1}}{\sqrt{2}} \otimes\frac{\ket{0}+\ket{1}}{\sqrt{2}}\otimes \frac{\ket{0}+\ket{1}}{\sqrt{2}}
\end{align*}
$$

$$
\begin{align*}
= \frac{1}{2\sqrt{2}}(|000\rangle+ |001\rangle+ |010\rangle+ |011\rangle+ |100\rangle+ |101\rangle+ |110\rangle+ |111\rangle)
\end{align*}
$$


Usando qiskit obtenemos el mismo resultado para el estado del sistema antes de las mediciones:

In [106]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>


Noten que si aplicamos un operador al estado cuántico $|\psi\rangle$ lo estamos aplicando simultaneamente a todos los estados

$|000\rangle$,$|001\rangle$, ...., $|111\rangle$.

Esto se llama **paralelismo cuántica** y es una de la ventajas de la computación cuántica.

Enseguida simulamos este circuito

In [ ]:
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1000)
print(job.result().get_counts())
plot_histogram(job.result().get_counts())

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{2.4 Uso de una Computadora Cuántica Real de IBM Quantum Experience}}$
<hr style="border:2px solid gray"> </hr>

Lo primero es registrarse en


https://quantum-computing.ibm.com/

Una vez que el email se verifica ya tendremos una CLAVE de acceso y la guardamos con

In [16]:
from qiskit import IBMQ
IBMQ.save_account('CLAVE', overwrite=True)


Para obtener la lista de computadoras reales disponibles usar el código siguiente:


In [111]:
from qiskit import IBMQ
IBMQ.load_account() # Load account from disk
print(IBMQ.providers())    # List all available providers
provider = IBMQ.get_provider(hub='ibm-q')
provider.backends()

ibmqfactory.load_account:WARNING:2023-06-16 23:00:21,326: Credentials are already in use. The existing account in the session will be replaced.


[<AccountProvider for IBMQ(hub='ibm-q', group='open', project='main')>]


[<IBMQSimulator('ibmq_qasm_simulator') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQBackend('ibmq_lima') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQBackend('ibmq_belem') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQBackend('ibmq_quito') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQSimulator('simulator_statevector') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQSimulator('simulator_mps') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQSimulator('simulator_extended_stabilizer') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQSimulator('simulator_stabilizer') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQBackend('ibmq_jakarta') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQBackend('ibmq_manila') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQBackend('ibm_lagos') from IBMQ(hub='ibm-q', group='open', project='main')>,
 <IBMQBackend('ibm_nairo

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{2.5 Cálculo de Valores de Expectación de Operadores}}$
<hr style="border:2px solid gray"> </hr>

Primero definimos el estado cuántico $|\psi\rangle$ y el operador usando **qiskit.opflow**.

Por ejemplo para calcular
$$
\langle\psi|X\otimes X\otimes X|\psi\rangle
$$

para
$$  |\psi\rangle =\frac{\ket{000}+\ket{111}}{\sqrt{2}}$$

lo podemos hacer del modo siguiente en Qiskit.




In [112]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp,Zero, One, Z, X,Y,I
import numpy as np
psi = 1 / np.sqrt(2) * ((One ^ One^ One) + (Zero ^ Zero^ Zero))
qc=QuantumCircuit(3)
qc.x(0)
qc.x(1)
qc.x(2)
op=CircuitOp(qc)
from IPython.display import display, Latex,Math
display(Math('|\psi> = \\frac{1}{\sqrt{2}}(|0>|0>|0>+|1>|1>|1>)'))
display(Latex(f'$<\psi|X\otimes X\otimes X|\psi>$'))
print("=",np.round(psi.adjoint().compose(op).compose(psi).eval().real,1))

<ipython-input-112-28d337920216>:9: DeprecationWarning: The class ``qiskit.opflow.primitive_ops.circuit_op.CircuitOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  op=CircuitOp(qc)


<IPython.core.display.Math object>

<IPython.core.display.Latex object>

= 1.0


O simplificarlo más en la definición del operador, como se muestra a continuación

In [113]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp,Zero, One, Z, X,Y,I
import numpy as np
psi = 1 / np.sqrt(2) * ((One ^ One^ One) + (Zero ^ Zero^ Zero))
op = X^X^X
from IPython.display import display, Latex,Math
display(Math('|\psi> = \\frac{1}{\sqrt{2}}(|0>|0>|0>+|1>|1>|1>)'))
display(Latex(f'$<\psi|Op|\psi>$'))
print("=",np.round(psi.adjoint().compose(op).compose(psi).eval().real,1))


<IPython.core.display.Math object>

<IPython.core.display.Latex object>

= 1.0


Por ejemplo para calcular
$$
\langle\psi|XX+YY+ZZ|\psi\rangle
$$

para
$$  |\psi\rangle =\ket{0}$$

lo podemos hacer del modo siguiente:


In [114]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp,Zero, One, Z, X,Y,I
import numpy as np
psi = Zero
from IPython.display import display, Latex,Math
display(Math('|\psi> = |0>'))
display(Latex(f'$<\psi|Op^\dag Op|\psi>$'))
print("=",np.round(psi.adjoint().compose(X.adjoint()).compose(X).compose(psi).eval().real,1))
print("=",np.round(psi.adjoint().compose(Y.adjoint()).compose(Y).compose(psi).eval().real,1))
print("=",np.round(psi.adjoint().compose(Z.adjoint()).compose(Z).compose(psi).eval().real,1))


<IPython.core.display.Math object>

<IPython.core.display.Latex object>

= 1.0
= 1.0
= 1.0


De modo que
$$
\langle\psi|XX+YY+ZZ|\psi\rangle = 3
$$

para
$$  |\psi\rangle =\ket{0}$$


Recuérdese que el operador de spin es
$$
S= \frac{\hbar}{2}(\hat iX+\hat jY+\hat kZ)
$$


Para un electrón es
$$
S^2 \mid s,m_z\rangle = s(s+1) \hbar^2 \mid s,m_z\rangle=\frac{3}{4}\hbar^2 \mid s,m_z\rangle
$$
con $s=\frac{1}{2}$



<hr style="border:2px solid gray"> </hr>

# $\color{blue}{\text{3. Entrelazamiento de 2 Qubits. Estados de Bell}}$
<hr style="border:2px solid gray"> </hr>

Los 3 estados de Bell son los siguientes:
$$
|\Phi^+\rangle = \frac{1}{\sqrt{2}} (|0\rangle \otimes |0\rangle + |1\rangle \otimes |1\rangle) \\
|\Phi^-\rangle = \frac{1}{\sqrt{2}} (|0\rangle \otimes |0\rangle - |1\rangle \otimes |1\rangle) \\
|\Psi^+\rangle = \frac{1}{\sqrt{2}} (|0\rangle \otimes |1\rangle + |1\rangle \otimes |0\rangle) \\
|\Psi^-\rangle = \frac{1}{\sqrt{2}} (|0\rangle \otimes |1\rangle - |1\rangle \otimes |0\rangle)
$$

o bien

$$
|\Phi^+\rangle = \frac{1}{\sqrt{2}} (|0\rangle |0\rangle + |1\rangle  |1\rangle) \\
|\Phi^-\rangle = \frac{1}{\sqrt{2}} (|0\rangle  |0\rangle - |1\rangle  |1\rangle) \\
|\Psi^+\rangle = \frac{1}{\sqrt{2}} (|0\rangle |1\rangle + |1\rangle  |0\rangle) \\
|\Psi^-\rangle = \frac{1}{\sqrt{2}} (|0\rangle  |1\rangle - |1\rangle  |0\rangle)
$$

Normalmente se usa la siguiente notación simplificada:

$$\newcommand{\ket}[1]{\left|{#1}\right\rangle}$$
$$\newcommand{\bra}[1]{\left\langle{#1}\right|}$$

$$  \ket{\Phi^+} =\frac{\ket{00}+\ket{11}}{\sqrt{2}}$$

$$  \ket{\Phi^-} =\frac{\ket{00}-\ket{11}}{\sqrt{2}}$$

$$  \ket{\Psi^+} =\frac{\ket{01}+\ket{10}}{\sqrt{2}}$$

$$  \ket{\Psi^-} =\frac{\ket{01}-\ket{10}}{\sqrt{2}}$$

Por ejemplo, para **crear el primer estado de Bell** se usa el circuito siguiente:

In [ ]:
from qiskit import QuantumCircuit, Aer,execute
import qiskit.quantum_info as qi
qc = QuantumCircuit(2,2) # Ponemos 2 qubits y 2 bits
# Aplicamos algunas compuertas
qc.h(0) # Hadamard gate
qc.cx(0,1) # CNOT gate
stv = qi.Statevector.from_instruction(qc)
for qubit in range(2):
  qc.measure(qubit,qubit)
qc.draw("mpl")

Las compuertas cuánticas aplicadas se expresan matemáticamente del modo siguiente:

$$
\begin{align*} (CNOT  )(I \otimes H  ) |00\rangle =(CNOT  )\ket{0}\otimes\frac{\ket{0}+\ket{1}}{\sqrt{2}} = \frac{1}{\sqrt{2}}(CNOT)( |00\rangle+|01\rangle) = \frac{1}{\sqrt{2}}(|00\rangle+|11\rangle)
\end{align*}
$$


Usando qiskit vemos que el estado del sistema antes de las mediciones es el mismo que acabamos de obtener:

---



In [31]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

Podemos usar el simulador simple en JavaScript

https://ciiec.buap.mx/QCSim/

Enseguida simulamos este circuito

In [ ]:
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1000)
print(job.result().get_counts())
plot_histogram(job.result().get_counts())

El circuito para crear el segundo estado de Bell es

In [ ]:
from qiskit import QuantumCircuit, Aer,execute
import qiskit.quantum_info as qi
qc = QuantumCircuit(2,2) # Ponemos 2 qubits y 2 bits
qc.h(0) # Hadamard gate
qc.cx(0,1) # CNOT gate
qc.z(1) # NOT gate
stv = qi.Statevector.from_instruction(qc)
for qubit in range(2):
  qc.measure(qubit,qubit)
qc.draw("mpl")

In [118]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

In [ ]:
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1000)
print(job.result().get_counts())
plot_histogram(job.result().get_counts())

    Usando


https://ciiec.buap.mx/QCSim/


    Bell state

    verbose 1;

    h q[0];

    cx q[0], q[1];

    z  q[1];

El circuito para crear el tercer estado de Bell es

In [ ]:
from qiskit import QuantumCircuit, Aer,execute
import qiskit.quantum_info as qi
qc = QuantumCircuit(2,2) # Ponemos 2 qubits y 2 bits
qc.h(0) # Hadamard gate
qc.cx(0,1) # CNOT gate
qc.x(1) # NOT gate
stv = qi.Statevector.from_instruction(qc)
for qubit in range(2):
  qc.measure(qubit,qubit)
qc.draw("mpl")

In [121]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

In [ ]:
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1000)
print(job.result().get_counts())
plot_histogram(job.result().get_counts())

Finalmente, el circuito para crear el cuarto estado de Bell es

In [ ]:
from qiskit import QuantumCircuit, Aer,execute
import qiskit.quantum_info as qi
qc = QuantumCircuit(2,2) # Ponemos 2 qubits y 2 bits
qc.h(0) # Hadamard gate
qc.cx(0,1) # CNOT gate
qc.x(1) # NOT gate
qc.z(1) # Z gate
qc.barrier()
stv = qi.Statevector.from_instruction(qc)
for qubit in range(2):
  qc.measure(qubit,qubit)
qc.draw("mpl")

Nótese que podemos intercambiar las compuertas X y Z y obtener el mismo resultado (salvo una fase global).

In [124]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

In [ ]:
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1000)
print(job.result().get_counts())
plot_histogram(job.result().get_counts())


<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{3.1 Ejemplo de Suma de Espines para 2 Electrones con Qiskit}}$
<hr style="border:2px solid gray"> </hr>

Por ejemplo, suponiendo que los qubits representen el estado de espín de un electrón, podemos calcular el espín total 2 electrones dependiendo del estado cuántico.

Para los 3 estados triplete
$$  |\psi\rangle =\frac{\ket{00}+\ket{11}}{\sqrt{2}}$$
$$  |\psi\rangle =\frac{\ket{00}-\ket{11}}{\sqrt{2}}$$
$$  |\psi\rangle =\frac{\ket{01}+\ket{10}}{\sqrt{2}}$$

el resultado es
$$
\langle\psi|(S_1+S_2)^2|\psi\rangle = \langle\psi|(S_1^2+S_2^2+2S_1\cdot S_2)|\psi\rangle=8 \frac{\hbar^2}{4} = 2 \hbar^2
$$


y para el estado singlete
$$  |\psi\rangle =\frac{\ket{01}-\ket{10}}{\sqrt{2}}$$

$$
\langle\psi|(S_1+S_2)^2|\psi\rangle = \langle\psi|(S_1^2+S_2^2+2S_1\cdot S_2)|\psi\rangle=0
$$


Podemos hacer de forma sencilla estos cálculos en Qiskit del modo siguiente:


In [126]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp,Zero, One, Z, X,Y,I
import numpy as np
from IPython.display import display, Latex,Math
Op1 = X^X
Op2 = Y^Y
Op3 = Z^Z
for j in range(4):
  if j == 0:
    psi = 1 / np.sqrt(2) * ((Zero^ Zero) + (One^ One))
    display(Math('|\psi> = \\frac{1}{\sqrt{2}}(|0>|0>+|1>|1>)'))
  if j == 1:
    psi = 1 / np.sqrt(2) * ((Zero^ Zero) - (One^ One))
    display(Math('|\psi> = \\frac{1}{\sqrt{2}}(|0>|0>-|1>|1>)'))
  if j == 2:
    psi = 1 / np.sqrt(2) * ((Zero^ One) + (One^ Zero))
    display(Math('|\psi> = \\frac{1}{\sqrt{2}}(|0>|1>+|1>|0>)'))
  if j == 3:
    psi = 1 / np.sqrt(2) * ((Zero^ One) - (One^ Zero))
    display(Math('|\psi> = \\frac{1}{\sqrt{2}}(|0>|1>-|1>|0>)'))
  display(Latex(f'$<\psi|2(X\otimes X+Y\otimes Y+Z\otimes Z)|\psi>$'))

  print("=",2*(np.round(psi.adjoint().compose(Op1).compose(psi).eval().real,1)
  +np.round(psi.adjoint().compose(Op2).compose(psi).eval().real,1)+
  np.round(psi.adjoint().compose(Op3).compose(psi).eval().real,1)),"\n")


<IPython.core.display.Math object>

<IPython.core.display.Latex object>

= 2.0 



<IPython.core.display.Math object>

<IPython.core.display.Latex object>

= 2.0 



<IPython.core.display.Math object>

<IPython.core.display.Latex object>

= 2.0 



<IPython.core.display.Math object>

<IPython.core.display.Latex object>

= -6.0 



De modo que el spin total para los estados triplete de 2 electrones es 1, es decir que

$$
S^2 \mid s,m_z\rangle = s(s+1) \hbar^2 \mid s,m_z\rangle = 2 \hbar^2
$$
con $s=1$

y 0 para el estado singlete

$$
S^2 \mid s,m_z\rangle = s(s+1) \hbar^2 \mid s,m_z\rangle = 0
$$
con $s=0$

<hr style="border:2px solid gray"> </hr>

# $\color{blue}{\text{4. Entrelazamiento de 3 Qubits. Estado GHZ}}$
<hr style="border:2px solid gray"> </hr>

Para 3 qubits los estados equivalentes a los de Bell de máximo entrelazamiento se llaman estados de **Greenberger–Horne–Zeilinger**.

Tienen propiedades muy interesantes. Un ejemplo es

$$\newcommand{\ket}[1]{\left|{#1}\right\rangle}$$
$$\newcommand{\bra}[1]{\left\langle{#1}\right|}$$

$$  \ket{GHZ} =\frac{\ket{000}+\ket{111}}{\sqrt{2}}$$

In [ ]:
from qiskit import QuantumCircuit, Aer,transpile
import qiskit.quantum_info as qi
qc = QuantumCircuit(3,3) # Ponemos 3 qubits y 3 bits
# Aplicamos algunas compuertas
qc.h(0) # Hadamard gate
qc.cx(0,1) # CNOT gate
qc.cx(1,2) # CNOT gate
stv = qi.Statevector.from_instruction(qc)
for qubit in range(3):
  qc.measure(qubit,qubit)
qc.draw("mpl")

Las compuertas cuánticas aplicadas se expresan matemáticamente del modo siguiente:

$$
\begin{align*} (CNOT \otimes I)(I \otimes CNOT)(I \otimes I \otimes H ) |000\rangle =\frac{1}{\sqrt{2}} (CNOT \otimes I)(I \otimes CNOT)(\ket{000}+ \ket{001}) = \frac{1}{\sqrt{2}}(CNOT \otimes I)(\ket{000}+ \ket{011}) = \frac{1}{\sqrt{2}}(\ket{000}+ \ket{111})
\end{align*}
$$


El estado del sistema antes de las mediciones obtenido mediante qiskit es el mismo:

In [128]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

Enseguida simulamos este circuito

In [ ]:
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1000)
print(job.result().get_counts())
plot_histogram(job.result().get_counts())

Nuevamente podemos usar el simulador simple en JavaScript

https://ciiec.buap.mx/QCSim/

con el código siguiente:

h q[0] ;# Hadamard gate

cx q[0],q[1]; # CNOT gate

cx q[1],q[2]; # CNOT gate


Una versión en Python de este simulador simple está en

https://github.com/lvillasen/Quantum-Computer-Simulator

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{4.1 Estado GHZ con Ruido Simulado Acorde con una Computadora Real}}$
<hr style="border:2px solid gray"> </hr>

Basta con apuntar la variable **backend** a alguna computadora real específica de IBM Quantum Experience con
```
IBMQ.save_account('poner-la-clave-de IBMQ', overwrite=True)
provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
```
o a la más desocupada con

```
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
      and not x.configuration().simulator and x.status().operational==True))
```
y correr

```
noise_model = NoiseModel.from_backend(backend)
print("----------Noise Model for",backend,"--------------\n",noise_model)
sim_backend = AerSimulator.from_backend(backend)
tqc = transpile(qc, sim_backend,optimization_level=3)
job = sim_backend.run(tqc,shots=1000)
print("Counts for 3-qubit GHZ state with simulated noise model for", backend)
plot_histogram(job.result().get_counts())
```

In [129]:
from qiskit import IBMQ, transpile
from qiskit import QuantumCircuit
from qiskit_aer import AerSimulator
from qiskit.tools.visualization import plot_histogram
from qiskit.providers.ibmq import least_busy


Escogemos la computadora **ibmq_manila**.

In [130]:
from qiskit import IBMQ, Aer
from qiskit_aer.noise import NoiseModel
#IBMQ.save_account('CLAVE', overwrite=True)

provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
noise_model = NoiseModel.from_backend(backend)
print("----------Noise Model for",backend,"--------------\n",noise_model)

ibmqfactory.load_account:WARNING:2023-06-16 23:26:54,072: Credentials are already in use. The existing account in the session will be replaced.


----------Noise Model for ibmq_manila --------------
 NoiseModel:
  Basis gates: ['cx', 'id', 'reset', 'rz', 'sx', 'x']
  Instructions with noise: ['sx', 'measure', 'cx', 'reset', 'id', 'x']
  Qubits with noise: [0, 1, 2, 3, 4]
  Specific qubit errors: [('id', (0,)), ('id', (1,)), ('id', (2,)), ('id', (3,)), ('id', (4,)), ('sx', (0,)), ('sx', (1,)), ('sx', (2,)), ('sx', (3,)), ('sx', (4,)), ('x', (0,)), ('x', (1,)), ('x', (2,)), ('x', (3,)), ('x', (4,)), ('cx', (4, 3)), ('cx', (3, 4)), ('cx', (2, 3)), ('cx', (3, 2)), ('cx', (1, 2)), ('cx', (2, 1)), ('cx', (0, 1)), ('cx', (1, 0)), ('reset', (0,)), ('reset', (1,)), ('reset', (2,)), ('reset', (3,)), ('reset', (4,)), ('measure', (0,)), ('measure', (1,)), ('measure', (2,)), ('measure', (3,)), ('measure', (4,))]


O escogemos la más desocupada.

In [ ]:
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
      and not x.configuration().simulator and x.status().operational==True))
print("least busy backend: ", backend)
noise_model = NoiseModel.from_backend(backend)
print("----------Noise Model for",backend,"--------------\n",noise_model)

In [ ]:
sim_backend = AerSimulator.from_backend(backend)
print("least busy backend: ", sim_backend)

# Transpile the circuit for the noisy basis gates
tqc = transpile(qc, sim_backend,optimization_level=3)

# Execute noisy simulation and get counts
job = sim_backend.run(tqc,shots=1000)
print("Counts for 3-qubit GHZ state with simulated noise model for", backend)
plot_histogram(job.result().get_counts())

El diagrama modificado para esta computadora se llama circuito "transpilado"

In [ ]:
tqc.draw("mpl")

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{4.2 Estado GHZ en una Computadora Real}}$
<hr style="border:2px solid gray"> </hr>

Basta con apuntar la variable **backend** a alguna computadora real específica de IBM Quantum Experience con
```
IBMQ.save_account('poner-la-clave-de IBMQ', overwrite=True)
provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
```
o a la más desocupada con
```
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
      and not x.configuration().simulator and x.status().operational==True))
```
y correr
```
tqc = transpile(qc, backend,optimization_level=3)
job = backend.run(tqc,shots=1000)
```
Después monitorear la ejecución con
```
from qiskit.providers.jobstatus import JobStatus
print(job.status(),job.queue_position())
```
y acceder a los resultados una vez que termina con
```
print("Counts for 3-qubit GHZ state with real computer", backend)
plot_histogram(job.result().get_counts())
```


In [ ]:
from qiskit import QuantumCircuit, Aer,transpile
import qiskit.quantum_info as qi
qc = QuantumCircuit(3,3) # Ponemos 3 qubits y 3 bits
# Aplicamos algunas compuertas
qc.h(0) # Hadamard gate
qc.cx(0,1) # CNOT gate
qc.cx(1,2) # CNOT gate
stv = qi.Statevector.from_instruction(qc)
for qubit in range(3):
  qc.measure(qubit,qubit)
qc.draw("mpl")

In [135]:
from qiskit import IBMQ
from qiskit.providers.ibmq import least_busy
# Load local account information
IBMQ.load_account()
# Get the least busy backend
provider = IBMQ.get_provider(hub='ibm-q')
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
                                       and not x.configuration().simulator
                                       and x.status().operational==True))
print("least busy backend: ", backend)
# Run our circuit
tqc = transpile(qc, backend, optimization_level=3)
job = backend.run(tqc,shots = 100)

ibmqfactory.load_account:WARNING:2023-06-16 23:29:51,254: Credentials are already in use. The existing account in the session will be replaced.


least busy backend:  ibmq_belem


In [137]:
from qiskit.providers.jobstatus import JobStatus
print(job.status(),job.queue_position())

JobStatus.QUEUED None


In [ ]:
print("Counts for 3-qubit GHZ state with real computer", backend)
plot_histogram(job.result().get_counts())

<hr style="border:2px solid gray"> </hr>

# $\color{blue}{\text{5. Teleportación Cuántica}}$
<hr style="border:2px solid gray"> </hr>



A la fecha se ha logrado teletransportar qubits a distancias mayores a 1000 Km a través de un satélite Chino.

Vamos a suponer que Alice tiene un qubit arbitrario que quiere compartir a Bob. En Computación Cuántica es imposible hacer una copia de un qubit (teorema de no clonación), sin embargo se lo puede compartir a Bob a costa de destruir su qubit siguiendo el protocolo siguiente:

- Suponemos que el qubit que Alice quiere compartir es

$$ |\psi \rangle = \alpha|0\rangle + \beta|1\rangle $$

- Alice y Bob le piden a Charly que cree un par de qubits entrelazados

- Charly de da uno de esos 2 qubits entrelazados a Alice y el otro a Bob.

- Para propósitos del circuito cuántico que haremos, vamos a llamar $q_0$ a $ |\psi \rangle $, $q_1$ al qubit entrelazado de Alice y $q_2$ al qubit entrelazado de Bob

- Alice aplica una compuerta CNOT a su par de qubits $q_1q_0$

- Alice aplica una compuerta Hadamard a su qubit $q_0$

- Alice mide sus dos qubits y le pasa los resultados $C_1C_0$ a Bob por un canal clásico (con velocidad < c)

- Dependiendo de los valores $C_1C_0$ Bob aplica las operaciones cuánticas siguientes a su qubit:

<div align="center">Si $C_1C_0 = 00$ no hace nada </div>

<div align="center">Si $C_1C_0 = 01$ aplica una compuerta Z </div>

<div align="center">Si $C_1C_0 = 10$ aplica una compuerta X </div>

<div align="center">Si $C_1C_0 = 11$ aplica las compuertas ZX </div>

A continuación vamos a programar este protocolo para simularlo. También lo podemos correr en alguna de las computadores cuánticas de IBM con alguna variante para corroborar que sí se llevó a cabo con éxito ya que no se puede conocer el estado inicial de $ |\psi \rangle$ ni el estado final de $q_2$ para compararlos(al medirlos se destruyen).





El qubit arbitrario que Alice quiere teleportar a Bob es el siguiente:

In [3]:
import numpy as np
from qiskit.visualization import array_to_latex
from qiskit.quantum_info import random_statevector
psi = random_statevector(2)
display(array_to_latex(psi, prefix="|\\psi\\rangle ="))

<IPython.core.display.Latex object>

Su representación en la esfera de Bloch es


$$ |\psi \rangle = cos\frac{\theta}{2}|0\rangle + e^{i\phi}sin\frac{\theta}{2}|1\rangle $$


In [ ]:
from qiskit.visualization import plot_bloch_multivector
plot_bloch_multivector(psi)

Aplicando el protocolo llegamos al circuito cuántico siguiente:

In [ ]:
from qiskit import QuantumCircuit, assemble, Aer
import qiskit.quantum_info as qi
import numpy as np
import qiskit.quantum_info as qi
qc = QuantumCircuit(3,2)
qc.initialize(psi, 0)
stv0 = qi.Statevector.from_instruction(qc)
qc.barrier()
qc.h(1) # Hadamard
qc.cx(1,2) # CNOT
qc.barrier()
qc.cx(0,1)
qc.h(0)
qc.barrier()
stv = qi.Statevector.from_instruction(qc)
qc.measure(0,0)
qc.measure(1,1)
qc.draw("mpl")


El estado de los 3 qubits al inicio es el siguiente.

Como ya dijimos, los 2 primeros qubits ($q_1q_0$) los tiene Alice y el tercero ($q_2$) lo tiene Bob.

In [ ]:
from qiskit.visualization import plot_bloch_multivector
plot_bloch_multivector(stv0)

El estado del sistema justo antes de las mediciones de $q_0$ y de $q_1$ por Alice es:

In [7]:
stv.draw('latex', prefix='Estado \quad del \quad sistema \quad de \quad 3 \quad qubits \quad = \quad')

<IPython.core.display.Latex object>

Ahora Alice mide sus 2 qubits obteniendo el resultado siguiente:

In [ ]:
from qiskit import QuantumCircuit, Aer,execute
from qiskit.visualization import plot_histogram
backend = Aer.get_backend('qasm_simulator')
job = backend.run(qc, shots=1)
print(job.result().get_counts(qc))
plot_histogram(job.result().get_counts())

Dependiendo del resultado de los 2 bits que Alice le manda, Bob aplica las compuertas cuánticas a su qubit.


In [9]:
import qiskit.quantum_info as qi
import numpy as np
qc1 = QuantumCircuit(3,2)
print(job.result().get_counts())
for key in job.result().data()["counts"]:
  if int(key,0) == 0 :
    print("El resultado es 00")
    q2_state = [np.round(stv[0],3),np.round(stv[4],3)]
    qc1.initialize(0, 0)
    qc1.initialize(0, 1)
  elif int(key,0) == 1 :
    print("El resultado es 01")
    q2_state = [np.round(stv[1],3),np.round(stv[5],3)]
    qc1.initialize(1, 0)
    qc1.initialize(0, 1)
  elif int(key,0) == 2 :
    print("El resultado es 10")
    q2_state = [np.round(stv[2],3),np.round(stv[6],3)]
    qc1.initialize(0, 0)
    qc1.initialize(1, 1)
  elif int(key,0) == 3 :
    print("El resultado es 11")
    q2_state = [np.round(stv[3],3),np.round(stv[7],3)]
    qc1.initialize(1, 0)
    qc1.initialize(1, 1)
q2_normalizado = q2_state/np.linalg.norm(q2_state)
qc1.barrier()
qc1.initialize(q2_normalizado, 2)
for key in job.result().data()["counts"]:
  if int(key,0) == 1 :
    qc1.z(2)
  if int(key,0) == 2 :
    qc1.x(2)
  if int(key,0) == 3 :
    qc1.x(2)
    qc1.z(2)

stv1 = qi.Statevector.from_instruction(qc1)
stv1.draw('latex', prefix='Estado \quad del \quad qubit \quad 2 = ')

{'10': 1}
El resultado es 10


<IPython.core.display.Latex object>


Vemos que en efecto el qubit de Bob se transforma en el qubit $$ |\psi \rangle = \alpha|0\rangle + \beta|1\rangle $$ que Alice le teleportó.

El estado final de los 3 qubits es el siguiente:

In [ ]:
from qiskit.visualization import plot_bloch_multivector
plot_bloch_multivector(stv1)


El circuito cuántico completo de la teleportación en este caso es el siguiente:

In [ ]:
import qiskit.quantum_info as qi
import numpy as np
qc2 = QuantumCircuit(3,2)
for key in job.result().data(qc)["counts"]:
  if int(key,0) != 0 :
    qc2.initialize(psi, 2)
qcc = qc.compose(qc1)
qc3 = qcc.compose(qc2)
qc3.draw("mpl")

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{5.1 Sustento Matemático del Protocolo de Teleportación Cuántica}}$
<hr style="border:2px solid gray"> </hr>


- Suponemos que el qubit que Alice quiere compartir es

$$ \alpha|0\rangle + \beta|1\rangle $$

- Alice y Bob le piden a Charly que cree un par de qubits entrelazados

$$  \frac{1}{\sqrt{2}} (|00\rangle + |11\rangle) $$

- Charly le da uno de esos 2 qubits entrelazados (el de la derecha) a Alice y el otro a Bob


- El estado cuántico de los 3 qubits se obtiene haciendo el producto tensorial en el que Alice posee los dos primeros qubits y Bob el tercero.

$$ \begin{align*}
\frac{1}{\sqrt{2}} (|00\rangle + |11\rangle)\otimes ( \alpha|0\rangle+ \beta |1\rangle \\
&= \frac{1}{\sqrt{2}} (\alpha|000\rangle + \alpha|110\rangle + \beta|001\rangle + \beta|111\rangle)
\end{align*}$$

- Recordemos que

$$ CNOT \ket{00}= \ket{00}$$
$$ CNOT \ket{01}= \ket{11}$$
$$ CNOT \ket{10}= \ket{10}$$
$$ CNOT \ket{11}= \ket{01}$$


- Siguiendo el protocolo Alice aplica una compuerta CNOT a su par de qubits  $𝑞_1𝑞_0$ y enseguida una compuerta Hadamard a su qubit  $𝑞_0$.

El resultado del estado cuántico de los 3 qubits es el siguiente:

$$
\begin{align*} (I \otimes I \otimes H) (I \otimes CNOT) \frac{1}{\sqrt{2}} (\alpha|000\rangle + \alpha|110\rangle + \beta|001\rangle + \beta|111\rangle) \\
&= (I \otimes I \otimes H) \frac{1}{\sqrt{2}} (\alpha|000\rangle + \alpha|110\rangle + \beta|011\rangle + \beta|101\rangle) \\
&= \frac{1}{2}  (\alpha(|000\rangle + |001\rangle + |110\rangle + |111\rangle) + \beta(|010\rangle - |011\rangle + |100\rangle - |101\rangle)) \\
\end{align*}
$$

que se puede reescribir como

$$
\begin{align*}
= \frac{1}{2}( (\alpha|0\rangle + \beta|1\rangle)\otimes|00\rangle
+ (\alpha|1\rangle - \beta|0\rangle)\otimes|01\rangle   + (\alpha|1\rangle + \beta|0\rangle)\otimes|10\rangle
 + (\alpha|1\rangle - \beta|0\rangle)\otimes|11\rangle  )
\end{align*}
$$

- Enseguida Alice mide sus 2 qubits y le manda el resultado de los 2 bits a través de un canal clásico

Alice obtiene como resultado uno de los cuatro posibles resultados  $|00\rangle, |01\rangle, |10\rangle,$ o $|11\rangle$ con la misma probabilidad.

El qubit de Bob se proyectará en uno de los 4 estados siguientes:

$$ |00\rangle \rightarrow (\alpha|0\rangle + \beta|1\rangle)$$

$$|01\rangle \rightarrow (\alpha|1\rangle - \beta|0\rangle)$$

$$|10\rangle \rightarrow (\alpha|1\rangle + \beta|0\rangle)$$

$$|11\rangle \rightarrow (\alpha|1\rangle - \beta|0\rangle)$$

- Recordemos que

$$ X \ket{0} = \ket{1}$$
$$ X \ket{1} = \ket{0}$$

y


$$ Z \ket{0}= \ket{0} $$
$$ Z \ket{1}= -\ket{1} $$


- Dependiendo de los valores $C_1C_0$ Bob aplica las operaciones cuánticas siguientes a su qubit:


<div align="center">Si $C_1C_0 = 00$ Bob no hace nada </div>

<div align="center">Si $C_1C_0 = 01$ Bob aplica una compuerta Z a su qubit</div>

<div align="center">Si $C_1C_0 = 10$ Bob aplica una compuerta X a su qubit</div>

<div align="center">Si $C_1C_0 = 11$ Bob aplica las compuertas ZX a su qubit </div>

- Al final el estado del qubit de Bob será

<div align="center"> $|q_2\rangle= \alpha|0\rangle + \beta|1\rangle$ </div>

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{5.2 Teleportación Cuántica en una Computadora Real}}$
<hr style="border:2px solid gray"> </hr>

El siguiente código se encuentra también en

https://github.com/lvillasen/Quantum-Teleportation-with-Qiskit-on-a-Real-Computer/blob/main/Quantum-Teleportation.ipynb

Apuntamos la variable **backend** a alguna computadora real específica de IBM Quantum Experience con
```
IBMQ.save_account('poner-la-clave-de IBMQ', overwrite=True)
provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
```
o a la más desocupada con
```
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
      and not x.configuration().simulator and x.status().operational==True))
```
hacemos la transpilación del circuito y lo ejecutamos
```
tqc = transpile(qc, backend,optimization_level=3)
job = backend.run(tqc,shots=1000)
```
Después monitoreamos la ejecución con
```
from qiskit.providers.jobstatus import JobStatus
print(job.status(),job.queue_position())
```
y accedemos a los resultados una vez que termina con
```
print("Counts for 3-qubit GHZ state with real computer", backend)
plot_histogram(job.result().get_counts())
```
Dado que en una computadora real no se permite por el momento dar más instrucciones después de alguna medición, vamos a adaptar el código de la teleportación general que vimos en la sección anterior para teleportar solamente cuando Alice mide $\ket{00}$ en sus dos qubit e ignoramos los otros 3 casos.

El circuito siguiente incluye la generación aleatoria mediante los ángulos $\theta$ y $\phi$ del qubit que se desea teleportar.

$$
|\psi\rangle = \cos({\frac{\theta}{2}})|0\rangle +e^{i\phi} \sin({\frac{\theta}{2}})|1\rangle=\begin{pmatrix} \cos({\frac{\theta}{2}}) \\\\ e^{i\phi}\sin({\frac{\theta}{2}}) \end{pmatrix}
$$

vemos que los operadores unitarios de rotación


$$R_Y(\theta) = \exp\left(-i \frac{\theta}{2} Y\right) =
    \begin{pmatrix}
        \cos{\frac{\theta}{2}} & -\sin{\frac{\theta}{2}} \\
        \sin{\frac{\theta}{2}} & \cos{\frac{\theta}{2}}
    \end{pmatrix}
$$

y

$$R_Z(\phi) = \exp\left(-i \frac{\phi}{2} Z\right) =
    \begin{pmatrix}
        e^{-i\phi/2} & 0 \\
       0 & e^{i\phi/2}
    \end{pmatrix}
$$



nos permiten transformar $|0\rangle$ en  $|\psi\rangle$ (hasta un factor de fase) como se muestra a continuación
$$
R_Z(\phi) R_Y(\theta) |0\rangle =\begin{pmatrix}
        e^{-i\phi/2} & 0 \\
       0 & e^{i\phi/2}
    \end{pmatrix}
\begin{pmatrix}
        \cos{\frac{\theta}{2}} & -\sin{\frac{\theta}{2}} \\
        \sin{\frac{\theta}{2}} & \cos{\frac{\theta}{2}}
\end{pmatrix}
    \begin{pmatrix}1 \\\\ 0 \end{pmatrix} =
    \begin{pmatrix} e^{-i\phi/2}\cos{\frac{\theta}{2}} \\\\ e^{i\phi/2}\sin{\frac{\theta}{2}}  \end{pmatrix}
    =e^{-i\phi/2}|\psi\rangle
$$



In [ ]:
from qiskit import QuantumCircuit, assemble, Aer
from qiskit import IBMQ, transpile
from qiskit.providers.ibmq import least_busy
from qiskit.tools.visualization import plot_histogram

import qiskit.quantum_info as qi
import numpy as np
import qiskit.quantum_info as qi
theta=np.pi*np.random.rand()
phi = 2*np.pi*np.random.rand()
print("theta =",np.round(theta,2),"rad,   theta =",np.round(180*theta/np.pi,2),"grados")
print("phi=",np.round(phi,2),"rad,   phi =",np.round(180*phi/np.pi,2),"grados")
qc = QuantumCircuit(3,3)
qc.ry(theta,0)
qc.rz(phi,0)
stv0 = qi.Statevector.from_instruction(qc)
qc.barrier()
qc.h(1) # Hadamard
qc.cx(1,2) # CNOT
qc.barrier()
qc.cx(0,1)
qc.h(0)
qc.barrier()

qc.rz(-phi,2)
qc.ry(-theta,2)
qc.measure(0,0)
qc.measure(1,1)
qc.measure(2,2)
qc.draw("mpl")

Los 3 qubits en la primera barrera son

In [ ]:
from qiskit.visualization import plot_bloch_multivector
plot_bloch_multivector(stv0)

Primero simulamos el circuito

In [ ]:
backend = Aer.get_backend('qasm_simulator')
#job = backend.run(qc,backend, shots=1000)
job = backend.run(qc, shots=1000)

#print(job.result().get_counts(qc))
try :
  n_error = job.result().get_counts(qc)['100']
except:
  n_error = 0
import matplotlib.pyplot as plt
fig = plt.figure(figsize=(10,6))

states = ['000', '100']
counts = [job.result().get_counts(qc)['000'],n_error]
plt.rcParams.update({'font.size': 20})
plt.bar(states,counts,width=0.4, bottom=None,align='center', data=None)
for index, value in enumerate(counts):
    plt.text(index, value, str(value),fontsize=20)
plt.xlabel('States',fontsize=20)
plt.ylabel('Counts',fontsize=20)
plt.show()

#plot_histogram(job.result().get_counts())

Vemos que en el caso que nos interesa, que es cuando los 2 primeros qubits están en $\ket{00}$, el tercer qubit siempre está en $\ket{0}$, es decir que el estado $\ket{100}$ no aparece en la medición, de modo que la teleportación de $\ket{\psi}$ es perfecta ya que los dos operadores de rotación aplicados al final siempre transforman  $|\psi\rangle$ en
$|0\rangle$ (salvo una fase)
$$
R_Y(-\theta) R_Z(-\phi)|\psi\rangle =\begin{pmatrix}
        \cos{\frac{\theta}{2}} & \sin{\frac{\theta}{2}} \\
        -\sin{\frac{\theta}{2}} & \cos{\frac{\theta}{2}}
    \end{pmatrix}
    \begin{pmatrix}
        e^{i\phi/2} & 0 \\
       0 & e^{-i\phi/2}
    \end{pmatrix}
    \begin{pmatrix} \cos({\frac{\theta}{2}}) \\\\ e^{i\phi}\sin({\frac{\theta}{2}}) \end{pmatrix} =e^{i\phi/2}
    \begin{pmatrix} 1 \\\\ 0 \end{pmatrix}
    =e^{i\phi/2}|0\rangle
$$

Ahora lo ejecutamos en una computadora con ruido similar al de una computadora real

In [ ]:
from qiskit_aer import AerSimulator

IBMQ.load_account()
# Get the least busy backend
provider = IBMQ.get_provider(hub='ibm-q')
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
                                       and not x.configuration().simulator
                                       and x.status().operational==True))
sim_backend = AerSimulator.from_backend(backend)
print("least busy backend: ", sim_backend,"\n\n")

# Transpile the circuit for the noisy basis gates
tqc = transpile(qc, sim_backend,optimization_level=3)

# Execute noisy simulation and get counts
job = sim_backend.run(tqc,shots=1000)
try :
  n_error = job.result().get_counts(qc)['100']
except:
  n_error = 0
import matplotlib.pyplot as plt
try :
  n_error = job.result().get_counts(qc)['100']
except:
  n_error = 0
import matplotlib.pyplot as plt
fig = plt.figure(figsize=(10,6))

states = ['000', '100']
counts = [job.result().get_counts(qc)['000'],n_error]
plt.rcParams.update({'font.size': 20})
plt.bar(states,counts,width=0.4, bottom=None,align='center', data=None)
for index, value in enumerate(counts):
    plt.text(index, value, str(value),fontsize=20)
plt.xlabel('States',fontsize=20)
plt.ylabel('Counts',fontsize=20)
plt.show()



Vemos que la teleportación ya no es perfecta como lo evidencia la presencia del estado $|100\rangle$ en la medición.

El diagrama transpilado es el siguiente

In [ ]:
tqc.draw("mpl")

Ahora lo ejecutamos en una computadora real

In [19]:
from qiskit import QuantumCircuit, Aer,execute
from qiskit.visualization import plot_histogram
IBMQ.load_account()
# Get the least busy backend
provider = IBMQ.get_provider(hub='ibm-q')
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
                                       and not x.configuration().simulator
                                       and x.status().operational==True))
print("least busy backend: ", backend)
tqc = transpile(qc, backend, optimization_level=3)
job_teleportation = backend.run(tqc,shots = 1000)


ibmqfactory.load_account:WARNING:2023-06-17 16:00:37,247: Credentials are already in use. The existing account in the session will be replaced.


least busy backend:  ibmq_lima


Monitoreamos su avance en la fila con

In [20]:
print(job_teleportation.status(),job_teleportation.queue_position())

JobStatus.QUEUED None


Al terminar accedemos a los datos con

In [ ]:
print(job_teleportation.result().get_counts(),"\n\n")
import matplotlib.pyplot as plt
fig = plt.figure(figsize=(10,8))
states = ['000', '100']
counts = [job.result().get_counts(qc)['000'],n_error]
plt.rcParams.update({'font.size': 20})
plt.bar(states,counts,width=0.4, bottom=None,align='center', data=None)
for index, value in enumerate(counts):
    plt.text(index, value, str(value),fontsize=20)
plt.xlabel('States',fontsize=20)
plt.ylabel('Counts',fontsize=20)
plt.show()

Vemos que hay errores en la teleportación cuántica del orden del 5%.


<hr style="border:2px solid gray"> </hr>

# $\color{blue}{\text{6. Desigualdad de Bell}}$
<hr style="border:2px solid gray"> </hr>

Esta parte está basada en la variante de desigualdad de Bell propuesta por David Mermin en 1981.

El código está también en

https://github.com/lvillasen/Testing-Bell-Inequality-on-a-Quantum-Computer

Ver por ejemplo

https://www.physicsforums.com/insights/answering-mermins-challenge-with-wilczeks-challenge/

Ver el video de Brian Greene

https://www.youtube.com/watch?v=UZiwtfrisTQ


<img src="https://www.physicsforums.com/insights/wp-content/uploads/2019/10/MerminDevice-1500x627.png" alt="Alternative text" />



<img src="https://www.physicsforums.com/insights/wp-content/uploads/2019/10/EPRBmeasurements-1500x848.png" alt="Alternative text" />



Fuente: https://www.physicsforums.com/insights/answering-mermins-challenge-with-wilczeks-challenge/

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{6.1 Predicción de EPR}}$
<hr style="border:2px solid gray"> </hr>

```
EXPERIMENTO EPR vs MECANICA CUANTICA

Premio Nobel de Física 2022 al francés Alain Aspect, al estadounidense John Clauser y al austriaco Anton Zeilinger

Medimos el espín de las partículas A y B en 3 ejes que estám de 0 (Eje 1), 120 (Eje 2) y 240 grados (Eje 3)



El espín medido 0 corresponde a spin up y el 1 a spin down a lo largo del eje respectivo


Medición A <---------------------------- Par A B entrelazados con S = 0 -------------------> Medición B



Medimos el espín A y B en ángulos de 0, 120 0 240 grados


Plan A  Plan B	 Anti-Correlaciones		 		Fracción Anticorrelacionados

123	  123	 [A1B1,A1B2,A1B3,A2B1,A2B2,A2B3,A3B1,A3B2,A3B3]

000 	 111 	 [1, 1, 1, 1, 1, 1, 1, 1, 1] 			 9 / 9
001 	 110 	 [1, 1, 0, 1, 1, 0, 0, 0, 1] 			 5 / 9
010 	 101 	 [1, 0, 1, 0, 1, 0, 1, 0, 1] 			 5 / 9
011 	 100 	 [1, 0, 0, 0, 1, 1, 0, 1, 1] 			 5 / 9
100 	 011 	 [1, 0, 0, 0, 1, 1, 0, 1, 1] 			 5 / 9
101 	 010 	 [1, 0, 1, 0, 1, 0, 1, 0, 1] 			 5 / 9
110 	 001 	 [1, 1, 0, 1, 1, 0, 0, 0, 1] 			 5 / 9
111 	 000 	 [1, 1, 1, 1, 1, 1, 1, 1, 1] 			 9 / 9


```
De acuerdo con EPR, es decir con una teoría local de variables ocultas, la fracción de anticorrelaciones debe ser > 5/9 (la desigualdad de Bell es AntiCorr(A,B) > 56%) pero el experimento arroja 50%

Es decir que la MC viola la desigualdad de Bell y por lo tanto la visión de EPR no coincide con el experimento (la MC es no local)



<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{6.2 Predicción de la Mecánica Cuántica}}$
<hr style="border:2px solid gray"> </hr>


Para medir la proyección del operador de espín en una dirección arbitraria

$$\hat{n} = \hat{i} sin\theta cos\phi  + \hat{j} sin\theta sin\phi + \hat{k} cos \theta$$

usamos

$$\hat{\sigma} \cdot\hat{n}$$

con

$$\hat{\sigma} = \hat{i}X + \hat{j}Y+\hat{k}Z$$

Por ejemplo si aplicamos el operador $\hat{\sigma} \cdot\hat{n}$ al estado cuántico

$$
|\psi\rangle = \cos{\frac{\theta}{2}}|0\rangle + e^{i\phi}\sin{\frac{\theta}{2}}|1\rangle
$$
con $0\leq \phi < 2\pi$, y $0\leq \theta \leq \pi$.

obtenemos
$$\hat{\sigma} \cdot\hat{n}|\psi\rangle = (n_xX+n_yY+n_zZ)|\psi\rangle$$

Recordando que los operadores de Pauli se pueden escribir como

$$ X =\ket{0}\bra{1} + \ket{1}\bra{0}$$
$$ Y =-i\ket{0}\bra{1} + i\ket{1}\bra{0}$$
$$ Z =\ket{0}\bra{0} - \ket{1}\bra{1}$$

$$\hat{\sigma} \cdot\hat{n}|\psi\rangle =\left(n_x\left(\ket{0}\bra{1} + \ket{1}\bra{0}\right)+n_y\left(-i\ket{0}\bra{1} + i\ket{1}\bra{0}\right)+n_z\left(\ket{0}\bra{0} - \ket{1}\bra{1}\right)\right)\left(\cos{\frac{\theta}{2}}|0\rangle + e^{i\phi}\sin{\frac{\theta}{2}}|1\rangle\right)$$

Haciendo los productos internos obtenemos

$$\hat{\sigma} \cdot\hat{n}|\psi\rangle= \left((n_x-in_y)e^{i\phi}\sin{\frac{\theta}{2}}+n_z\cos{\frac{\theta}{2}}\right)\ket{0}+
\left((n_x+in_y)\cos{\frac{\theta}{2}}-n_ze^{i\phi}\sin{\frac{\theta}{2}}\right)\ket{1}$$

sustituyendo

$$n_x = sin\theta cos\phi$$
$$n_y = sin\theta sin\phi$$
$$n_z = cos\theta $$

obtenemos

$$(n_x-in_y)e^{i\phi}\sin{\frac{\theta}{2}}+n_z\cos{\frac{\theta}{2}}=\sin\theta\sin{\frac{\theta}{2}}+\cos\theta\cos{\frac{\theta}{2}}= \sin^2\frac{\theta}{2}\cos{\frac{\theta}{2}}+ \cos^3{\frac{\theta}{2}}= \cos{\frac{\theta}{2}}$$
y
$$\left((n_x+in_y)\cos{\frac{\theta}{2}}-n_ze^{i\phi}\sin{\frac{\theta}{2}}\right)=e^{i\phi}(\sin\theta\cos{\frac{\theta}{2}}-\cos\theta\sin{\frac{\theta}{2}})=e^{i\phi}\sin{\frac{\theta}{2}}$$

De modo que en efecto $$\hat{\sigma} \cdot\hat{n}|\psi\rangle= |\psi\rangle$$

Igualmente para el estado cuántico $|\psi ^{'}\rangle$, que en la esfera de Block apunta en sentido contrario a $|\psi\rangle$, dado por

$$
|\psi^{'}\rangle = \cos{\frac{\pi-\theta}{2}}|0\rangle + e^{i(\phi+\pi)}\sin{\frac{\pi-\theta}{2}}|1\rangle=\sin{\frac{\theta}{2}}|0\rangle - e^{i\phi}\cos{\frac{\theta}{2}}|1\rangle
$$

vemos que $|\psi^{'}\rangle$ es también eigenvector de $\hat{\sigma} \cdot\hat{n}$ pero con eigenvalor $-1$:


$$\hat{\sigma} \cdot\hat{n}|\psi^{'}\rangle =\left(n_x\left(\ket{0}\bra{1} + \ket{1}\bra{0}\right)+n_y\left(-i\ket{0}\bra{1} + i\ket{1}\bra{0}\right)+n_z\left(\ket{0}\bra{0} - \ket{1}\bra{1}\right)\right)\left(\sin{\frac{\theta}{2}}|0\rangle - e^{i\phi}\cos{\frac{\theta}{2}}|1\rangle\right)$$

Haciendo los productos internos obtenemos

$$\hat{\sigma} \cdot\hat{n}|\psi^{'}\rangle= \left(-(n_x-in_y)e^{i\phi}\cos{\frac{\theta}{2}}+n_z\sin{\frac{\theta}{2}}\right)\ket{0}+
\left((n_x+in_y)\sin{\frac{\theta}{2}}+n_ze^{i\phi}\cos{\frac{\theta}{2}}\right)\ket{1}$$

sustituyendo

$$n_x = sin\theta cos\phi$$
$$n_y = sin\theta sin\phi$$
$$n_z = cos\theta $$

obtenemos

$$-(n_x-in_y)e^{i\phi}\cos{\frac{\theta}{2}}+n_z\sin{\frac{\theta}{2}}=-\sin\theta\cos{\frac{\theta}{2}}+\cos\theta\sin{\frac{\theta}{2}}= -\cos^2\frac{\theta}{2}\sin{\frac{\theta}{2}}- \sin^3{\frac{\theta}{2}}= -\sin{\frac{\theta}{2}}$$
y
$$\left((n_x+in_y)\sin{\frac{\theta}{2}}+n_ze^{i\phi}\cos{\frac{\theta}{2}}\right)=e^{i\phi}(\sin\theta\sin{\frac{\theta}{2}}+\cos\theta\cos{\frac{\theta}{2}})=e^{i\phi}\cos{\frac{\theta}{2}}$$

De modo que en efecto

$$\hat{\sigma} \cdot\hat{n}|\psi^{'}\rangle= -\sin{\frac{\theta}{2}}|0\rangle + e^{i\phi}\cos{\frac{\theta}{2}}|1\rangle$$

$$\hat{\sigma} \cdot\hat{n}|\psi^{'}\rangle= -|\psi^{'}\rangle$$


Los dos estados $|\psi\rangle$ y $|\psi^{'}\rangle$ son ortogonales ya que

$$\langle\psi|\psi^{'}\rangle = \left(\cos{\frac{\theta}{2}}\langle0| + e^{-i\phi}\sin{\frac{\theta}{2}}\langle1|\right)\left(\sin{\frac{\theta}{2}}|0\rangle - e^{i\phi}\cos{\frac{\theta}{2}}|1\rangle\right)=0$$


Alternativamente a la base computacional, podemos usar la base cruzada con $\theta=\frac{2\pi}{3}$ para medir la partícula A o B. Los 2 estados que conforman esta base son

$$|0\rangle_{120} = \cos(\frac{\pi}{3})|0\rangle+\sin(\frac{\pi}{3})|1\rangle$$
y
$$|1\rangle_{120} = \sin(\frac{\pi}{3})|0\rangle-\cos(\frac{\pi}{3})|1\rangle$$

Despejando $|0\rangle$ y $|1\rangle$ obtenemos
$$|0\rangle = \cos(\frac{\pi}{3})|0\rangle_{120}+\sin(\frac{\pi}{3})|1\rangle_{120}$$
y
$$|1\rangle = \sin(\frac{\pi}{3})|0\rangle_{120}-\cos(\frac{\pi}{3})|1\rangle_{120}$$




De modo que podemos escribir el estado singlete de Bell como

$$
|\Psi^-\rangle = \frac{1}{\sqrt{2}} (|0\rangle \otimes |1\rangle - |1\rangle \otimes |0\rangle)=
$$
$$\frac{1}{\sqrt{2}}(\cos(\frac{\pi}{3})|0\rangle_{120}+\sin(\frac{\pi}{3})|1\rangle_{120})\otimes(\sin(\frac{\pi}{3})|0\rangle_{120}-\cos(\frac{\pi}{3})|1\rangle_{120})
$$
$$-\frac{1}{\sqrt{2}}(\sin(\frac{\pi}{3})|0\rangle_{120}-\cos(\frac{\pi}{3})|1\rangle_{120})\otimes(\cos(\frac{\pi}{3})|0\rangle_{120}+\sin(\frac{\pi}{3})|1\rangle_{120})
$$

o sea
$$|\Psi^-\rangle =-\frac{1}{\sqrt{2}} (|0\rangle_{120} \otimes |1\rangle_{120} - |1\rangle_{120} \otimes |0\rangle_{120})$$

Cuando A y B se miden en los mismos ejes, es decir ambos verticales o ambos rotados igual 120 grados obtenemos que la probabilidad de obtener mediciones anticorrelacionadas

$$P(A=|0\rangle,B=|1\rangle)+P(A=|1\rangle,B=|0\rangle)=\frac{1}{2}+\frac{1}{2}=1$$

o

$$P(A=|0\rangle_{120},B=|1\rangle_{120})+P(A=|1\rangle_{120},B=|0\rangle_{120})=\frac{1}{2}+\frac{1}{2}=1$$

y la probabilidad de obtener mediciones correlacionadas es

$$P(A=|0\rangle,B=|0\rangle)+P(A=|1\rangle,B=|1\rangle)=0$$

o
$$P(A=|0\rangle_{120},B=|0\rangle_{120})+P(A=|1\rangle_{120},B=|1\rangle_{120})=0$$


Igualmente podemos expresar $|\Psi^-\rangle$ en la base mixta en la que medimos A en el eje vertical y B en un eje rotado 120 grados en el plano XZ
$$
|\Psi^-\rangle = \frac{1}{\sqrt{2}}\left(|0\rangle \otimes |1\rangle - |1\rangle \otimes |0\rangle\right)
$$
$$=\frac{1}{\sqrt{2}} \left(|0\rangle \otimes (\sin(\frac{\pi}{3})|0\rangle_{120}-\cos(\frac{\pi}{3})|1\rangle_{120}) - |1\rangle \otimes (\cos(\frac{\pi}{3})|0\rangle_{120}+\sin(\frac{\pi}{3})|1\rangle_{120})\right)
$$

que resulta en
$$
|\Psi^-\rangle =\frac{1}{\sqrt{2}} \left(\sin(\frac{\pi}{3})|0\rangle \otimes |0\rangle_{120}-\cos(\frac{\pi}{3})|0\rangle \otimes |1\rangle_{120} - \cos(\frac{\pi}{3})|1\rangle \otimes |0\rangle_{120}-\sin(\frac{\pi}{3})|1\rangle \otimes|1\rangle_{120})\right)
$$

**De esta última representación de $|\Psi^-\rangle$ obtenemos la predicción de la MC en el caso en que el espín de A se mide en el eje vertical y el de B en un eje rotado 120 grados en el plano XZ**.

La probabilidad de obtener mediciones anticorrelacionadas es

$$P(A=|0\rangle,B=|1\rangle_{120})+P(A=|1\rangle,B=|0\rangle_{120})=\cos^2(\frac{\pi}{3})= \frac{1}{4}$$

y la probabilidad de obtener mediciones correlacionadas es

$$P(A=|0\rangle,B=|0\rangle_{120})+P(A=|1\rangle,B=|1\rangle_{120})=\sin^2(\frac{\pi}{3})= \frac{3}{4}$$


En consecuencia, si corremos 1000 veces cada uno de los 9 experimento posibles (3*3 combinaciones de los ejes a 120 grados en el plano XZ) la Mecánica Cuántica predice que,  en promedio, obtendremos los resultados siguientes:
```
Experimento             Anticorrelaciones          Correlaciones

A!B1                    1000                       0
A1B2                     250                       750
A1B3                     250                       750
A2B1                     250                       750
A2B2                    1000                       0
A2B3                     250                       750
A3B1                     250                       750
A3B2                     250                       750
A3B3                    1000                       0


TOTAL                   4500                       4500

```

Vemos que la predicción de la Mecánica Cuántica:

Anticorrelaciones = 50%

viola la desigualdad de Bell, que en esta variante es:

Anticorrelaciones > 56%.


**Como sabemos desde los 70´s, el experimento coincide con la Mecánica Cuántica y por lo tanto la visión de EPR se descarta (la Mecánica Cuántica es no local).**

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{6.3 Prueba de Bell en una Computadora Cuántica Simulada Ideal}}$
<hr style="border:2px solid gray"> </hr>

Para ilustrar cómo se puede hacer el experimento de Bell en una computadora cuántica requerimos entender cómo hacer una medición en bases rotadas a ángulos arbitrarios $\theta$ y $\phi$, en forma similar a lo que hicimos para $\theta=\frac{2\pi}{3}$ y $\phi=0$.

Partiendo de los 2 estados cuánticos ortogonales
$$
|\psi\rangle = \cos({\frac{\theta}{2}})|0\rangle +e^{i\phi} \sin({\frac{\theta}{2}})|1\rangle=\begin{pmatrix} \cos({\frac{\theta}{2}}) \\\\ e^{i\phi}\sin({\frac{\theta}{2}}) \end{pmatrix}
$$
y

$$
|\psi^{'}\rangle =\cos({\frac{\pi-\theta}{2}})|0\rangle +e^{i(\pi+\phi)} \sin({\frac{\pi-\theta}{2}})|1\rangle=\begin{pmatrix} \sin({\frac{\theta}{2}}) \\\\ -e^{i\phi}\cos({\frac{\theta}{2}}) \end{pmatrix}
$$

vemos que los operadores unitarios de rotación


$$R_Y(\theta) = \exp\left(-i \frac{\theta}{2} Y\right) =
    \begin{pmatrix}
        \cos{\frac{\theta}{2}} & -\sin{\frac{\theta}{2}} \\
        \sin{\frac{\theta}{2}} & \cos{\frac{\theta}{2}}
    \end{pmatrix}
$$

y

$$R_Z(\phi) = \exp\left(-i \frac{\phi}{2} Z\right) =
    \begin{pmatrix}
        e^{-i\phi/2} & 0 \\
       0 & e^{i\phi/2}
    \end{pmatrix}
$$



transforman la base  $|\psi\rangle, |\psi^{'}\rangle$ en la base computacional
$|0\rangle, |1\rangle$ (salvo una fase)
$$
R_Y(-\theta) R_Z(-\phi)|\psi\rangle =\begin{pmatrix}
        \cos{\frac{\theta}{2}} & \sin{\frac{\theta}{2}} \\
        -\sin{\frac{\theta}{2}} & \cos{\frac{\theta}{2}}
    \end{pmatrix}
    \begin{pmatrix}
        e^{i\phi/2} & 0 \\
       0 & e^{-i\phi/2}
    \end{pmatrix}
    \begin{pmatrix} \cos({\frac{\theta}{2}}) \\\\ e^{i\phi}\sin({\frac{\theta}{2}}) \end{pmatrix} =e^{i\phi/2}
    \begin{pmatrix} 1 \\\\ 0 \end{pmatrix}
    =e^{i\phi/2}|0\rangle
$$
$$
R_Y(-\theta) R_Z(-\phi)|\psi^{'}\rangle =\begin{pmatrix}
        \cos{\frac{\theta}{2}} & \sin{\frac{\theta}{2}} \\
        -\sin{\frac{\theta}{2}} & \cos{\frac{\theta}{2}}
    \end{pmatrix}
    \begin{pmatrix}
        e^{i\phi/2} & 0 \\
       0 & e^{-i\phi/2}
    \end{pmatrix}
    \begin{pmatrix} \sin({\frac{\theta}{2}}) \\\\ -e^{i\phi}\cos({\frac{\theta}{2}}) \end{pmatrix}  =e^{i\phi/2}
    \begin{pmatrix} 0 \\\\ 1 \end{pmatrix}
    =e^{i\phi/2}|1\rangle
$$

De modo que para medir un qubit en la base $\{|\psi\rangle, |\psi^{'}\rangle\}$ basta con aplicar $R_Y(-\theta) R_Z(-\phi)$ a ese qubit justo antes de medirlo en la base computacional $\{|0\rangle, |1\rangle\}$ de Qiskit.

Con estas herramientas ya podemos programar el experimento de la desigualdad de Bell que vimos en la sección anterior.

Primero generamos el estado singlete de Bell, que, como vimos en la **Sección 3.1**, tiene espín total igual a 0 y por lo tanto si A y B se miden en el mismo eje arbitrario siempre se obtienen resultados anti-correlacionados.

In [ ]:
from qiskit import QuantumCircuit, Aer,execute
import qiskit.quantum_info as qi
import numpy as np
qc = QuantumCircuit(2,2) # Ponemos 2 qubits y 2 bits
qc.h(0)
qc.cx(0,1)
qc.x(1)
qc.z(1)
qc.barrier()
stv = qi.Statevector.from_instruction(qc)
qc.draw("mpl")


In [23]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

A continuación corremos los 9 experimentos

In [ ]:
backend = Aer.get_backend('qasm_simulator')

N_corr = 0
N_anticorr = 0
qc3 = QuantumCircuit(2, 2)
qc3.measure(0,0)
qc3.measure(1,1)
print("                       Experimento A1B1")
qc11=qc
print(qc11)
qc11=qc.compose(qc3)
job11 = execute(qc11,backend, shots=1000)
result = job11.result()
print("Resultados :",result.get_counts(qc11))
if job11.result().get_counts().get('11') is not None:
  N_corr += job11.result().get_counts()['11']
if job11.result().get_counts().get('00') is not None:
  N_corr += job11.result().get_counts()['00']
N_anticorr += job11.result().get_counts()['01']+job11.result().get_counts()['10']

print()
print("                       Experimento A1B2")
qc12 = QuantumCircuit(2, 2)
qc12.rz(0,1)
qc12.ry(-2*np.pi/3,1)
qc12.barrier()
qc12=qc.compose(qc12)
qc12=qc12.compose(qc3)
print(qc12)
job12 = execute(qc12,backend, shots=1000)
result = job12.result()
print("Resultados :",result.get_counts(qc12))
N_corr += job12.result().get_counts()['11']+job12.result().get_counts()['00']
N_anticorr += job12.result().get_counts()['01']+job12.result().get_counts()['10']

print()
print("                       Experimento A1B3")
qc13 = QuantumCircuit(2, 2)
qc13.rz(-np.pi,1)
qc13.ry(-2*np.pi/3,1)
qc13.barrier()
qc13=qc.compose(qc13)
qc13=qc13.compose(qc3)
print(qc13)
job13 = execute(qc13,backend, shots=1000)
result = job13.result()
print("Resultados :",result.get_counts(qc13))
N_corr += job13.result().get_counts()['11']+job13.result().get_counts()['00']
N_anticorr += job13.result().get_counts()['01']+job13.result().get_counts()['10']

print()
print("                       Experimento A2B1")
qc21 = QuantumCircuit(2, 2)
qc21.rz(0,0)
qc21.ry(-2*np.pi/3,0)
qc21.barrier()
qc21=qc.compose(qc21)
qc21=qc21.compose(qc3)
print(qc21)
job21 = execute(qc21,backend, shots=1000)
result = job21.result()
print("Resultados :",result.get_counts(qc21))
N_corr += job21.result().get_counts()['11']+job21.result().get_counts()['00']
N_anticorr += job21.result().get_counts()['01']+job21.result().get_counts()['10']

print()
print("                       Experimento A2B2")
qc22 = QuantumCircuit(2, 2)
qc22.rz(0,0)
qc22.ry(-2*np.pi/3,0)
qc22.rz(0,1)
qc22.ry(-2*np.pi/3,1)
qc22.barrier()
qc22=qc.compose(qc22)
qc22=qc22.compose(qc3)
print(qc22)
job22 = execute(qc22,backend, shots=1000)
result = job22.result()
print("Resultados :",result.get_counts(qc22))
if job22.result().get_counts().get('11') is not None:
  N_corr += job22.result().get_counts()['11']
if job22.result().get_counts().get('00') is not None:
  N_corr += job22.result().get_counts()['00']
N_anticorr += job22.result().get_counts()['01']+job22.result().get_counts()['10']

print()
print("                       Experimento A2B3")
qc23 = QuantumCircuit(2, 2)
qc23.rz(0,0)
qc23.ry(-2*np.pi/3,0)
qc23.rz(-np.pi,1)
qc23.ry(-2*np.pi/3,1)
qc23.barrier()
qc23=qc.compose(qc23)
qc23=qc23.compose(qc3)
print(qc23)
job23 = execute(qc23,backend, shots=1000)
result = job23.result()
print("Resultados :",result.get_counts(qc23))
N_corr += job23.result().get_counts()['11']+job23.result().get_counts()['00']
N_anticorr += job23.result().get_counts()['01']+job23.result().get_counts()['10']

print()
print("                       Experimento A3B1")
qc31 = QuantumCircuit(2, 2)
qc31.rz(-np.pi,0)
qc31.ry(-2*np.pi/3,0)
qc31.barrier()
qc31=qc.compose(qc31)
qc31=qc31.compose(qc3)
print(qc31)
job31 = execute(qc31,backend, shots=1000)
result = job31.result()
print("Resultados :",result.get_counts(qc31))
N_corr += job31.result().get_counts()['11']+job31.result().get_counts()['00']
N_anticorr += job31.result().get_counts()['01']+job31.result().get_counts()['10']

print()
print("                       Experimento A3B2")
qc32 = QuantumCircuit(2, 2)
qc32.rz(-np.pi,0)
qc32.ry(-2*np.pi/3,0)
qc32.rz(0,1)
qc32.ry(-2*np.pi/3,1)
qc32.barrier()
qc32=qc.compose(qc32)
qc32=qc32.compose(qc3)
print(qc32)
job32 = execute(qc32,backend, shots=1000)
result = job31.result()
print("Resultados :",result.get_counts(qc32))
N_corr += job32.result().get_counts()['11']+job32.result().get_counts()['00']
N_anticorr += job32.result().get_counts()['01']+job32.result().get_counts()['10']

print()
print("                       Experimento A3B3")
qc33 = QuantumCircuit(2, 2)
qc33.rz(-np.pi,0)
qc33.ry(-2*np.pi/3,0)
qc33.rz(-np.pi,1)
qc33.ry(-2*np.pi/3,1)
qc33.barrier()
qc33=qc.compose(qc33)
qc33=qc33.compose(qc3)
print(qc33)
job33 = execute(qc33,backend, shots=1000)
result = job33.result()
print("Resultados :",result.get_counts(qc33))
if job33.result().get_counts().get('11') is not None:
  N_corr += job33.result().get_counts()['11']
if job33.result().get_counts().get('00') is not None:
  N_corr += job33.result().get_counts()['00']
N_anticorr += job33.result().get_counts()['01']+job33.result().get_counts()['10']
print("Total correlated:", N_corr, " Total anti-correlated:", N_anticorr)

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{6.4 Prueba de Bell en una Computadora Cuántica Simulada con Ruido}}$
<hr style="border:2px solid gray"> </hr>

We can add a noise model obtained from a real quantum computer from the IBM Quantum Experience, for example with
```
IBMQ.save_account('IBMQ-password', overwrite=True)
provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
```

In [26]:
from qiskit import IBMQ, Aer
from qiskit_aer.noise import NoiseModel
provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
noise_model = NoiseModel.from_backend(backend)
print("----------Noise Model for",backend,"--------------\n",noise_model)

ibmqfactory.load_account:WARNING:2023-06-17 16:05:02,299: Credentials are already in use. The existing account in the session will be replaced.


----------Noise Model for ibmq_manila --------------
 NoiseModel:
  Basis gates: ['cx', 'id', 'reset', 'rz', 'sx', 'x']
  Instructions with noise: ['sx', 'id', 'measure', 'cx', 'reset', 'x']
  Qubits with noise: [0, 1, 2, 3, 4]
  Specific qubit errors: [('id', (0,)), ('id', (1,)), ('id', (2,)), ('id', (3,)), ('id', (4,)), ('sx', (0,)), ('sx', (1,)), ('sx', (2,)), ('sx', (3,)), ('sx', (4,)), ('x', (0,)), ('x', (1,)), ('x', (2,)), ('x', (3,)), ('x', (4,)), ('cx', (4, 3)), ('cx', (3, 4)), ('cx', (2, 3)), ('cx', (3, 2)), ('cx', (1, 2)), ('cx', (2, 1)), ('cx', (0, 1)), ('cx', (1, 0)), ('reset', (0,)), ('reset', (1,)), ('reset', (2,)), ('reset', (3,)), ('reset', (4,)), ('measure', (0,)), ('measure', (1,)), ('measure', (2,)), ('measure', (3,)), ('measure', (4,))]


or we can get the least busy quantum computer with the following code:

In [27]:
from qiskit import QuantumCircuit, Aer,execute
from qiskit.visualization import plot_histogram
IBMQ.load_account()
# Get the least busy backend
provider = IBMQ.get_provider(hub='ibm-q')
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
                                       and not x.configuration().simulator
                                       and x.status().operational==True))
print("least busy backend: ", backend)
print("----------Noise Model for",backend,"--------------\n",noise_model)

ibmqfactory.load_account:WARNING:2023-06-17 16:05:10,902: Credentials are already in use. The existing account in the session will be replaced.


least busy backend:  ibmq_lima
----------Noise Model for ibmq_lima --------------
 NoiseModel:
  Basis gates: ['cx', 'id', 'reset', 'rz', 'sx', 'x']
  Instructions with noise: ['sx', 'id', 'measure', 'cx', 'reset', 'x']
  Qubits with noise: [0, 1, 2, 3, 4]
  Specific qubit errors: [('id', (0,)), ('id', (1,)), ('id', (2,)), ('id', (3,)), ('id', (4,)), ('sx', (0,)), ('sx', (1,)), ('sx', (2,)), ('sx', (3,)), ('sx', (4,)), ('x', (0,)), ('x', (1,)), ('x', (2,)), ('x', (3,)), ('x', (4,)), ('cx', (4, 3)), ('cx', (3, 4)), ('cx', (2, 3)), ('cx', (3, 2)), ('cx', (1, 2)), ('cx', (2, 1)), ('cx', (0, 1)), ('cx', (1, 0)), ('reset', (0,)), ('reset', (1,)), ('reset', (2,)), ('reset', (3,)), ('reset', (4,)), ('measure', (0,)), ('measure', (1,)), ('measure', (2,)), ('measure', (3,)), ('measure', (4,))]


The following code adds the noise model taken from the chosen backend to the
simulation

In [ ]:
from qiskit import QuantumCircuit, Aer
import qiskit.quantum_info as qi
import numpy as np
qc = QuantumCircuit(2,2) # Ponemos 2 qubits y 2 bits
qc.h(0)
qc.cx(0,1)
qc.x(1)
qc.z(1)
qc.barrier()
stv = qi.Statevector.from_instruction(qc)
qc.draw("mpl")


In [86]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

In [87]:
backend = AerSimulator.from_backend(backend)
shots = 1000
N_corr = 0
N_anticorr = 0
qc3 = QuantumCircuit(2, 2)
qc3.measure(0,0)
qc3.measure(1,1)
print("                       Experiment A1B1")
qc11=qc
qc11=qc.compose(qc3)
print(qc11)
tqc11 = transpile(qc11, backend, optimization_level=3)
job11 = backend.run(tqc11,shots =shots)
print("Results :",job11.result().get_counts())
if job11.result().get_counts().get('11') is not None:
  N_corr += job11.result().get_counts()['11']
if job11.result().get_counts().get('00') is not None:
  N_corr += job11.result().get_counts()['00']
N_anticorr += job11.result().get_counts()['01']+job11.result().get_counts()['10']

print("\n                       Experiment A1B2")
qc12 = QuantumCircuit(2, 2)
qc12.rz(0,1)
qc12.ry(-2*np.pi/3,1)
qc12.barrier()
qc12=qc.compose(qc12)
qc12=qc12.compose(qc3)
print(qc12)
tqc12 = transpile(qc12, backend, optimization_level=3)
job12 = backend.run(tqc12,shots =shots)
print("Results :",job12.result().get_counts())
N_corr += job12.result().get_counts()['11']+job12.result().get_counts()['00']
N_anticorr += job12.result().get_counts()['01']+job12.result().get_counts()['10']

print("\n                       Experiment A1B3")
qc13 = QuantumCircuit(2, 2)
qc13.rz(-np.pi,1)
qc13.ry(-2*np.pi/3,1)
qc13.barrier()
qc13=qc.compose(qc13)
qc13=qc13.compose(qc3)
print(qc13)
tqc13 = transpile(qc13, backend, optimization_level=3)
job13 = backend.run(tqc13,shots =shots)
print("Results :",job13.result().get_counts())
N_corr += job13.result().get_counts()['11']+job13.result().get_counts()['00']
N_anticorr += job13.result().get_counts()['01']+job13.result().get_counts()['10']

print("\n                       Experiment A2B1")
qc21 = QuantumCircuit(2, 2)
qc21.rz(0,0)
qc21.ry(-2*np.pi/3,0)
qc21.barrier()
qc21=qc.compose(qc21)
qc21=qc21.compose(qc3)
print(qc21)
tqc21 = transpile(qc21, backend, optimization_level=3)
job21 = backend.run(tqc21,shots =shots)
print("Results :",job21.result().get_counts())
N_corr += job21.result().get_counts()['11']+job21.result().get_counts()['00']
N_anticorr += job21.result().get_counts()['01']+job21.result().get_counts()['10']

print("\n                        Experiment A2B2")
qc22 = QuantumCircuit(2, 2)
qc22.rz(0,0)
qc22.ry(-2*np.pi/3,0)
qc22.rz(0,1)
qc22.ry(-2*np.pi/3,1)
qc22.barrier()
qc22=qc.compose(qc22)
qc22=qc22.compose(qc3)
print(qc22)
tqc22 = transpile(qc22, backend, optimization_level=3)
job22 = backend.run(tqc22,shots =shots)
print("Results :",job22.result().get_counts())
if job22.result().get_counts().get('11') is not None:
  N_corr += job22.result().get_counts()['11']
if job22.result().get_counts().get('00') is not None:
  N_corr += job22.result().get_counts()['00']
N_anticorr += job22.result().get_counts()['01']+job22.result().get_counts()['10']

print("\n                        Experiment A2B3")
qc23 = QuantumCircuit(2, 2)
qc23.rz(0,0)
qc23.ry(-2*np.pi/3,0)
qc23.rz(-np.pi,1)
qc23.ry(-2*np.pi/3,1)
qc23.barrier()
qc23=qc.compose(qc23)
qc23=qc23.compose(qc3)
print(qc23)
tqc23 = transpile(qc23, backend, optimization_level=3)
job23 = backend.run(tqc23,shots =shots)
print("Results :",job23.result().get_counts())
N_corr += job23.result().get_counts()['11']+job23.result().get_counts()['00']
N_anticorr += job23.result().get_counts()['01']+job23.result().get_counts()['10']

print("\n                        Experiment A3B1")
qc31 = QuantumCircuit(2, 2)
qc31.rz(-np.pi,0)
qc31.ry(-2*np.pi/3,0)
qc31.barrier()
qc31=qc.compose(qc31)
qc31=qc31.compose(qc3)
print(qc31)
tqc31 = transpile(qc31, backend, optimization_level=3)
job31 = backend.run(tqc31,shots =shots)
print("Results :",job31.result().get_counts())
N_corr += job31.result().get_counts()['11']+job31.result().get_counts()['00']
N_anticorr += job31.result().get_counts()['01']+job31.result().get_counts()['10']

print("\n                        Experiment A3B2")
qc32 = QuantumCircuit(2, 2)
qc32.rz(-np.pi,0)
qc32.ry(-2*np.pi/3,0)
qc32.rz(0,1)
qc32.ry(-2*np.pi/3,1)
qc32.barrier()
qc32=qc.compose(qc32)
qc32=qc32.compose(qc3)
print(qc32)
tqc32 = transpile(qc32, backend, optimization_level=3)
job32 = backend.run(tqc32,shots =shots)
print("Results :",job32.result().get_counts(qc32))
N_corr += job32.result().get_counts()['11']+job32.result().get_counts()['00']
N_anticorr += job32.result().get_counts()['01']+job32.result().get_counts()['10']

print("\n                        Experiment A3B3")
qc33 = QuantumCircuit(2, 2)
qc33.rz(-np.pi,0)
qc33.ry(-2*np.pi/3,0)
qc33.rz(-np.pi,1)
qc33.ry(-2*np.pi/3,1)
qc33.barrier()
qc33=qc.compose(qc33)
qc33=qc33.compose(qc3)
print(qc33)
tqc33 = transpile(qc33, backend, optimization_level=3)
job33 = backend.run(tqc33,shots =shots)
print("Results :",job33.result().get_counts(qc33))
if job33.result().get_counts().get('11') is not None:
  N_corr += job33.result().get_counts()['11']
if job33.result().get_counts().get('00') is not None:
  N_corr += job33.result().get_counts()['00']
N_anticorr += job33.result().get_counts()['01']+job33.result().get_counts()['10']
print("\nOut of a total of",9*shots,"measurements, we obtained")
print("Total correlated measurements:", N_corr)
print("Total anti-correlated measurements:", N_anticorr," corresponding to ",np.round(100*N_anticorr/(9*shots),2),"%")


                       Experiment A1B1
     ┌───┐                ░ ┌─┐   
q_0: ┤ H ├──■─────────────░─┤M├───
     └───┘┌─┴─┐┌───┐┌───┐ ░ └╥┘┌─┐
q_1: ─────┤ X ├┤ X ├┤ Z ├─░──╫─┤M├
          └───┘└───┘└───┘ ░  ║ └╥┘
c: 2/════════════════════════╩══╩═
                             0  1 
Results : {'11': 8, '01': 484, '00': 40, '10': 468}

                       Experiment A1B2
     ┌───┐                ░                        ░ ┌─┐   
q_0: ┤ H ├──■─────────────░────────────────────────░─┤M├───
     └───┘┌─┴─┐┌───┐┌───┐ ░ ┌───────┐┌───────────┐ ░ └╥┘┌─┐
q_1: ─────┤ X ├┤ X ├┤ Z ├─░─┤ Rz(0) ├┤ Ry(-2π/3) ├─░──╫─┤M├
          └───┘└───┘└───┘ ░ └───────┘└───────────┘ ░  ║ └╥┘
c: 2/═════════════════════════════════════════════════╩══╩═
                                                      0  1 
Results : {'01': 136, '11': 364, '10': 134, '00': 366}

                       Experiment A1B3
     ┌───┐                ░                         ░ ┌─┐   
q_0: ┤ H ├──■─────────────░─────────────────────

<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{6.5 Prueba de Bell en una Computadora Cuántica Real}}$
<hr style="border:2px solid gray"> </hr>


We can select a specific real quantum computer from the IBM Quantum Experience, for example with
```
IBMQ.save_account('IBMQ-password', overwrite=True)
provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
```


In [ ]:
from qiskit import IBMQ, Aer
from qiskit_aer.noise import NoiseModel
IBMQ.save_account('your-IBMQ-password', overwrite=True)
provider = IBMQ.load_account()
backend = provider.get_backend('ibmq_manila')
noise_model = NoiseModel.from_backend(backend)
print("----------Noise Model for",backend,"--------------\n",noise_model)

configrc.store_credentials:WARNING:2023-02-21 16:13:36,477: Credentials already present. Set overwrite=True to overwrite.
ibmqfactory.load_account:WARNING:2023-02-21 16:13:36,681: Credentials are already in use. The existing account in the session will be replaced.


----------Noise Model for ibmq_manila --------------
 NoiseModel:
  Basis gates: ['cx', 'id', 'reset', 'rz', 'sx', 'x']
  Instructions with noise: ['cx', 'x', 'id', 'reset', 'measure', 'sx']
  Qubits with noise: [0, 1, 2, 3, 4]
  Specific qubit errors: [('id', (0,)), ('id', (1,)), ('id', (2,)), ('id', (3,)), ('id', (4,)), ('sx', (0,)), ('sx', (1,)), ('sx', (2,)), ('sx', (3,)), ('sx', (4,)), ('x', (0,)), ('x', (1,)), ('x', (2,)), ('x', (3,)), ('x', (4,)), ('cx', (4, 3)), ('cx', (3, 4)), ('cx', (2, 3)), ('cx', (3, 2)), ('cx', (1, 2)), ('cx', (2, 1)), ('cx', (0, 1)), ('cx', (1, 0)), ('reset', (0,)), ('reset', (1,)), ('reset', (2,)), ('reset', (3,)), ('reset', (4,)), ('measure', (0,)), ('measure', (1,)), ('measure', (2,)), ('measure', (3,)), ('measure', (4,))]


or we can get the least busy quantum computer with the following code:

In [ ]:
from qiskit import QuantumCircuit, Aer,execute
from qiskit.visualization import plot_histogram
IBMQ.load_account()
# Get the least busy backend
provider = IBMQ.get_provider(hub='ibm-q')
backend = least_busy(provider.backends(filters=lambda x: x.configuration().n_qubits >= 2
                                       and not x.configuration().simulator
                                       and x.status().operational==True))
print("least busy backend: ", backend)
print("----------Noise Model for",backend,"--------------\n",noise_model)

ibmqfactory.load_account:WARNING:2023-02-21 16:13:45,085: Credentials are already in use. The existing account in the session will be replaced.


least busy backend:  ibmq_belem
----------Noise Model for ibmq_belem --------------
 NoiseModel:
  Basis gates: ['cx', 'id', 'reset', 'rz', 'sx', 'x']
  Instructions with noise: ['cx', 'x', 'id', 'reset', 'measure', 'sx']
  Qubits with noise: [0, 1, 2, 3, 4]
  Specific qubit errors: [('id', (0,)), ('id', (1,)), ('id', (2,)), ('id', (3,)), ('id', (4,)), ('sx', (0,)), ('sx', (1,)), ('sx', (2,)), ('sx', (3,)), ('sx', (4,)), ('x', (0,)), ('x', (1,)), ('x', (2,)), ('x', (3,)), ('x', (4,)), ('cx', (4, 3)), ('cx', (3, 4)), ('cx', (2, 3)), ('cx', (3, 2)), ('cx', (1, 2)), ('cx', (2, 1)), ('cx', (0, 1)), ('cx', (1, 0)), ('reset', (0,)), ('reset', (1,)), ('reset', (2,)), ('reset', (3,)), ('reset', (4,)), ('measure', (0,)), ('measure', (1,)), ('measure', (2,)), ('measure', (3,)), ('measure', (4,))]


The base circuit is

In [ ]:
from qiskit import QuantumCircuit, Aer
import qiskit.quantum_info as qi
import numpy as np
qc = QuantumCircuit(2,2) # Ponemos 2 qubits y 2 bits
qc.h(0)
qc.cx(0,1)
qc.x(1)
qc.z(1)
qc.barrier()
stv = qi.Statevector.from_instruction(qc)
qc.draw("mpl")

In [ ]:
stv.draw('latex', prefix="|\\psi\\rangle =")

<IPython.core.display.Latex object>

We run the first 5 experiments with

In [ ]:
shots = 1000
N_corr = 0
N_anticorr = 0
qc3 = QuantumCircuit(2, 2)
qc3.measure(0,0)
qc3.measure(1,1)
print("                       Experiment A1B1")
qc11=qc
qc11=qc.compose(qc3)
print(qc11)
tqc11 = transpile(qc11, backend, optimization_level=3)
job11 = backend.run(tqc11,shots =shots)
print("\n                       Experiment A1B2")
qc12 = QuantumCircuit(2, 2)
qc12.rz(0,1)
qc12.ry(-2*np.pi/3,1)
qc12.barrier()
qc12=qc.compose(qc12)
qc12=qc12.compose(qc3)
print(qc12)
tqc12 = transpile(qc12, backend, optimization_level=3)
job12 = backend.run(tqc12,shots =shots)
print("\n                       Experiment A1B3")
qc13 = QuantumCircuit(2, 2)
qc13.rz(-np.pi,1)
qc13.ry(-2*np.pi/3,1)
qc13.barrier()
qc13=qc.compose(qc13)
qc13=qc13.compose(qc3)
print(qc13)
tqc13 = transpile(qc13, backend, optimization_level=3)
job13 = backend.run(tqc13,shots =shots)
print("\n                       Experiment A2B1")
qc21 = QuantumCircuit(2, 2)
qc21.rz(0,0)
qc21.ry(-2*np.pi/3,0)
qc21.barrier()
qc21=qc.compose(qc21)
qc21=qc21.compose(qc3)
print(qc21)
tqc21 = transpile(qc21, backend, optimization_level=3)
job21 = backend.run(tqc21,shots =shots)
print("\n                        Experiment A2B2")
qc22 = QuantumCircuit(2, 2)
qc22.rz(0,0)
qc22.ry(-2*np.pi/3,0)
qc22.rz(0,1)
qc22.ry(-2*np.pi/3,1)
qc22.barrier()
qc22=qc.compose(qc22)
qc22=qc22.compose(qc3)
print(qc22)
tqc22 = transpile(qc22, backend, optimization_level=3)
job22 = backend.run(tqc22,shots =shots)

We can monitor them with

In [ ]:
from qiskit.providers.jobstatus import JobStatus
print(job11.status(),job11.queue_position())
print(job12.status(),job12.queue_position())
print(job13.status(),job13.queue_position())
print(job21.status(),job21.queue_position())
print(job22.status(),job22.queue_position())


Once the first 5 jobs have run we send the last 4 jobs with

In [ ]:
print("\n                        Experiment A2B3")
qc23 = QuantumCircuit(2, 2)
qc23.rz(0,0)
qc23.ry(-2*np.pi/3,0)
qc23.rz(-np.pi,1)
qc23.ry(-2*np.pi/3,1)
qc23.barrier()
qc23=qc.compose(qc23)
qc23=qc23.compose(qc3)
print(qc23)
tqc23 = transpile(qc23, backend, optimization_level=3)
job23 = backend.run(tqc23,shots =shots)
print("\n                        Experiment A3B1")
qc31 = QuantumCircuit(2, 2)
qc31.rz(-np.pi,0)
qc31.ry(-2*np.pi/3,0)
qc31.barrier()
qc31=qc.compose(qc31)
qc31=qc31.compose(qc3)
print(qc31)
tqc31 = transpile(qc31, backend, optimization_level=3)
job31 = backend.run(tqc31,shots =shots)
print("\n                        Experiment A3B2")
qc32 = QuantumCircuit(2, 2)
qc32.rz(-np.pi,0)
qc32.ry(-2*np.pi/3,0)
qc32.rz(0,1)
qc32.ry(-2*np.pi/3,1)
qc32.barrier()
qc32=qc.compose(qc32)
qc32=qc32.compose(qc3)
print(qc32)
tqc32 = transpile(qc32, backend, optimization_level=3)
job32 = backend.run(tqc32,shots =shots)
print("\n                        Experiment A3B3")
qc33 = QuantumCircuit(2, 2)
qc33.rz(-np.pi,0)
qc33.ry(-2*np.pi/3,0)
qc33.rz(-np.pi,1)
qc33.ry(-2*np.pi/3,1)
qc33.barrier()
qc33=qc.compose(qc33)
qc33=qc33.compose(qc3)
print(qc33)
tqc33 = transpile(qc33, backend, optimization_level=3)
job33 = backend.run(tqc33,shots =shots)

and monitor their execution with

In [ ]:
print(job23.status(),job23.queue_position())
print(job31.status(),job31.queue_position())
print(job32.status(),job32.queue_position())
print(job33.status(),job33.queue_position())

Once the 9 jobs finish we get their results with

In [ ]:
N_corr = 0
N_anticorr = 0
print("                       Experiment A1B1")
print("Results :",job11.result().get_counts())
if job11.result().get_counts().get('11') is not None:
  N_corr += job11.result().get_counts()['11']
if job11.result().get_counts().get('00') is not None:
  N_corr += job11.result().get_counts()['00']
N_anticorr += job11.result().get_counts()['01']+job11.result().get_counts()['10']
print("                       Experiment A1B2")
print("Results :",job12.result().get_counts())
N_corr += job12.result().get_counts()['11']
N_corr += job12.result().get_counts()['00']
N_anticorr += job12.result().get_counts()['01']+job12.result().get_counts()['10']
print("                       Experiment A1B3")
print("Results :",job13.result().get_counts())
N_corr += job13.result().get_counts()['11']
N_corr += job13.result().get_counts()['00']
N_anticorr += job13.result().get_counts()['01']+job13.result().get_counts()['10']
print("                       Experiment A2B1")
print("Results :",job21.result().get_counts())
N_corr += job21.result().get_counts()['11']
N_corr += job21.result().get_counts()['00']
N_anticorr += job21.result().get_counts()['01']+job21.result().get_counts()['10']
print("                       Experiment A2B2")
print("Results :",job22.result().get_counts())
if job22.result().get_counts().get('11') is not None:
  N_corr += job22.result().get_counts()['11']
if job22.result().get_counts().get('00') is not None:
  N_corr += job22.result().get_counts()['00']
N_anticorr += job22.result().get_counts()['01']+job22.result().get_counts()['10']
print("                       Experiment A2B3")
print("Results :",job23.result().get_counts())
N_corr += job23.result().get_counts()['11']
N_corr += job23.result().get_counts()['00']
N_anticorr += job23.result().get_counts()['01']+job23.result().get_counts()['10']
print("                       Experiment A3B1")
print("Results :",job31.result().get_counts())
N_corr += job31.result().get_counts()['11']
N_corr += job31.result().get_counts()['00']
N_anticorr += job31.result().get_counts()['01']+job31.result().get_counts()['10']
print("                       Experiment A3B2")
print("Results :",job32.result().get_counts())
N_corr += job32.result().get_counts()['11']
N_corr += job32.result().get_counts()['00']
N_anticorr += job32.result().get_counts()['01']+job32.result().get_counts()['10']
print("                       Experiment A3B3")
print("Results :",job33.result().get_counts())
if job33.result().get_counts().get('11') is not None:
  N_corr += job33.result().get_counts()['11']
if job33.result().get_counts().get('00') is not None:
  N_corr += job33.result().get_counts()['00']
N_anticorr += job33.result().get_counts()['01']+job33.result().get_counts()['10']

print("\nOut of a total of",9*shots,"measurements, we obtained")
print("Total correlated measurements:", N_corr)
print("Total anti-correlated measurements:", N_anticorr," corresponding to ",np.round(100*N_anticorr/(9*shots),2),"%")

<hr style="border:2px solid gray"> </hr>

# $\color{blue}{\text{7. Prueba de GHZ (Greenberger–Horne–Zeilinger)}}$
<hr style="border:2px solid gray"> </hr>


<hr style="border:2px solid gray"> </hr>

## $\color{blue}{\text{7.1 Contradicción entre la MC y una Teoría Local de Variables Ocultas}}$
<hr style="border:2px solid gray"> </hr>

Ver por ejemplo:

https://arxiv.org/abs/0712.0921

https://www.nature.com/articles/s41534-021-00397-z

Como vimos antes, por ejemplo para calcular
$$
\langle\psi|X\otimes X\otimes X|\psi\rangle
$$

para el estado cuántico
$$  |\psi\rangle =\frac{\ket{000}+\ket{111}}{\sqrt{2}}$$

lo podemos hacer del modo siguiente en Qiskit.
Primero definimos $|\psi\rangle$ mediante un circuito de cuántico.

In [139]:
from qiskit.opflow import CircuitStateFn
import qiskit.quantum_info as qi
import numpy as np
psi=QuantumCircuit(3)
psi.h(0) # Hadamard gate
psi.cx(0,1) # CNOT gate
psi.cx(1,2) # CNOT gate
stv0 = qi.Statevector.from_instruction(psi)
psi=CircuitStateFn(psi)
stv0.draw('latex', prefix="|\\psi\\rangle =")

<ipython-input-139-aee3e6c6cf23>:9: DeprecationWarning: The class ``qiskit.opflow.state_fns.circuit_state_fn.CircuitStateFn`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  psi=CircuitStateFn(psi)


<IPython.core.display.Latex object>

Enseguida definimos el operador mediante otro circuito.

In [140]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp
qc=QuantumCircuit(3)
qc.x(0)
qc.x(1)
qc.x(2)
op=CircuitOp(qc)
from IPython.display import display, Latex
display(Latex(f'$<\psi|X\otimes X\otimes X|\psi>$'))
print("=",np.round(psi.adjoint().compose(op).compose(psi).eval().real,1))


<ipython-input-140-4462639f06c9>:7: DeprecationWarning: The class ``qiskit.opflow.primitive_ops.circuit_op.CircuitOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  op=CircuitOp(qc)


<IPython.core.display.Latex object>

= 1.0


De este modo calculamos que

$$
\langle\psi|X\otimes X\otimes X|\psi\rangle =1
\\
\langle\psi|X\otimes Y\otimes Y|\psi\rangle =-1
\\
\langle\psi|Y\otimes X\otimes Y|\psi\rangle =-1
\\
\langle\psi|Y\otimes Y\otimes X|\psi\rangle =-1
$$

para el estado $  |GHZ\rangle $
$$  |\psi\rangle =\frac{\ket{000}+\ket{111}}{\sqrt{2}}$$


In [141]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp
qc=QuantumCircuit(3)
qc.x(0)
qc.x(1)
qc.x(2)
op=CircuitOp(qc)
from IPython.display import display, Latex
display(Latex(f'$<\psi|X\otimes X\otimes X|\psi>$'))
print("=",np.round(psi.adjoint().compose(op).compose(psi).eval().real,1))

<ipython-input-141-4462639f06c9>:7: DeprecationWarning: The class ``qiskit.opflow.primitive_ops.circuit_op.CircuitOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  op=CircuitOp(qc)


<IPython.core.display.Latex object>

= 1.0


In [142]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp
qc=QuantumCircuit(3)
qc.y(0)
qc.y(1)
qc.x(2)
op=CircuitOp(qc)
from IPython.display import display, Latex
display(Latex(f'$<\psi|X\otimes Y\otimes Y|\psi>$'))
print("=",np.round(psi.adjoint().compose(op).compose(psi).eval().real,1))

<ipython-input-142-8e1c8b78ce5b>:7: DeprecationWarning: The class ``qiskit.opflow.primitive_ops.circuit_op.CircuitOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  op=CircuitOp(qc)


<IPython.core.display.Latex object>

= -1.0


In [143]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp
qc=QuantumCircuit(3)
qc.y(0)
qc.x(1)
qc.y(2)
op=CircuitOp(qc)
from IPython.display import display, Latex
display(Latex(f'$<\psi|Y\otimes X\otimes Y|\psi>$'))
print("=",np.round(psi.adjoint().compose(op).compose(psi).eval().real,1))

<ipython-input-143-c7030f153a01>:7: DeprecationWarning: The class ``qiskit.opflow.primitive_ops.circuit_op.CircuitOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  op=CircuitOp(qc)


<IPython.core.display.Latex object>

= -1.0


In [93]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp
qc=QuantumCircuit(3)
qc.x(0)
qc.y(1)
qc.y(2)
op=CircuitOp(qc)
from IPython.display import display, Latex
display(Latex(f'$<\psi|Y\otimes Y\otimes X|\psi>$'))
print("=",np.round(psi.adjoint().compose(op).compose(psi).eval().real,1))

<ipython-input-93-7045571690be>:7: DeprecationWarning: The class ``qiskit.opflow.primitive_ops.circuit_op.CircuitOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  op=CircuitOp(qc)


<IPython.core.display.Latex object>

= -1.0


Podemos simplificar aún más este procedimiento del modo siguiente:

In [144]:
from qiskit import QuantumCircuit
from qiskit.opflow import CircuitOp,Zero, One, Z, X,Y,I
import numpy as np
psi = 1 / np.sqrt(2) * ((One ^ One^ One) + (Zero ^ Zero^ Zero))
op = X^X^X
from IPython.display import display, Latex,Math
display(Math('|\psi> = \\frac{1}{\sqrt{2}}(|0>|0>|0>+|1>|1>|1>)'))
for i in range(4):
  if i==0: Op = X^X^X
  if i==1: Op = X^Y^Y
  if i==2: Op = Y^X^Y
  if i==3: Op = Y^Y^X
  print("\nOp =",Op)
  display(Latex(f'$<\psi|Op|\psi>$'))
  print("=",np.round(psi.adjoint().compose(Op).compose(psi).eval().real,1))


<IPython.core.display.Math object>


Op = XXX


<IPython.core.display.Latex object>

= 1.0

Op = XYY


<IPython.core.display.Latex object>

= -1.0

Op = YXY


<IPython.core.display.Latex object>

= -1.0

Op = YYX


<IPython.core.display.Latex object>

= -1.0


Es decir que el producto de los 4 valores esperados es -1.

Si suponemos que existe una teoría local de variables ocultas, de modo que los valores esperados de los operadores X,Y,Z al actuar sobre cada una de las 3 partículas $i=1,2,3$ tienen valores predeterminados $v^i=\pm1$, entonces el producto de los 4 valores esperados para

$X\otimes X\otimes X$,

$X\otimes Y\otimes Y$,

$Y\otimes X\otimes Y$,

y

$Y\otimes Y\otimes X$,

debería ser

$v_{3x}v_{2x}v_{1X}v_{3x}v_{2y}v_{1y}v_{3y}v_{2x}v_{1Y}v_{3Y}v_{2y}v_{1x}=(v_{3x})^2(v_{2x})^2(v_{1x})^2(v_{3y})^2(v_{2y})^2(v_{1y})^2=1$

Mientras que el resultado de la Mecánica Cuántica es, como ya vimos, -1.

Con esta prueba se descartan las teorías locales de variables ocultas sin necesidad de calcular correlaciones como en el experimento de la desigualdad de Bell que vimos antes.